{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Tutorial - Part 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='oo_python'></a>\n",
    "## Object Oriented Python\n",
    "\n",
    "---\n",
    "The Python you have been writing so far is in the form of a procedural language. This is a logical procedure that takes input data, processes it, and produces output data. Object oriented programming is a method of programming that is organised around objects rather than actions and data rather than logic. OO programming takes the approach that what is important are the objects we want to manipulate rather than the logic required to manipulate them. For example, objects range from human beings (described by name, address etc) to buildings or GUIs and apps. All of these have features that can be described and managed.\n",
    "OO programming is not suitable to all tasks but if used appropriately it can provide benefits as it can sometimes better model the task being undertaken.\n",
    "\n",
    "Python is not a truly object oriented language, as Java is, but even when written as a procedural language it uses object oriented aspects of the language. In an object oriented language an object is an instance of a class and can access the methods (functions) within that class.\n",
    "\n",
    "Python lists, dictionaries and strings are in effect objects. For this reason they have their own methods that can be called:\n",
    "\n",
    "    String:\t          \tline.rstrip()\t\t\tStrip whitespace\n",
    "    List:\t          \tdemolist.index(42)\t\tReturn element at index 42\n",
    "\n",
    "---\n",
    "<a id='defining_class'></a>\n",
    "## Defining a Class\n",
    "\n",
    "A class is defined by the keyword “class”, the name and a semi colon. All code within the class is then indented.\n",
    "\n",
    "In Python 2 the class definition should always inherit \"object\", which means the declaration should be:\n",
    "\n",
    "\tclass Name(object):\n",
    "\n",
    "In Python 3 this is not necessary and there are 3 equally valid methods to declare the class:\n",
    "\n",
    "\tclass Name(object):\n",
    "\tclass Name():\n",
    "\tclass Name:\n",
    "\n",
    "This tutorial will use the last option but all can be used.\n",
    "\n",
    "---\n",
    "<a id='the_self'></a>\n",
    "## The self\n",
    "\n",
    "Class methods have only one specific difference from ordinary functions - they must have an extra first name that has to be added to the beginning of the parameter list, but you do do not give a value for this parameter when you call the method, Python will provide it. This particular variable refers to the object itself, and by convention, it is given the name self. \n",
    " \n",
    "Although, you can give any name for this parameter, it is strongly recommended that you use the name self.\n",
    "\n",
    "---\n",
    "<a id='example_class'></a>\n",
    "## Example Class\n",
    "\n",
    "A very simple class is shown below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Hello: \n",
    "    def sayHi(self): \n",
    "        print ('Hello world') \n",
    "\n",
    "h = Hello() \n",
    "h.sayHi() \n",
    "\n",
    "# This short example can also be written as Hello().sayHi() \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When run the output from the program is simple:\n",
    "\n",
    "\tHello world\n",
    "\n",
    "<b>NOTE:</b> The sayHi method takes no parameters but still has the self in the function definition. \n",
    "\n",
    "The above example has the class and the calling code in the same file, which will work but ideally you want the classes to be separate so you can use them as required.\n",
    "\n",
    "To achieve this, store the class code in a file called “hello.py”. The class can then be imported from this file:\n",
    "\n",
    "\tfrom hello import Hello\t\n",
    "\n",
    "\th = Hello() \n",
    "\th.sayHi() \n",
    "\n",
    "\n",
    "The class file can contain multiple classes, which can either be imported individually or all together with \"from hello import *\".\n",
    "\n",
    "The import method for classes is the same as for modules.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 24\n",
    "\n",
    "Write the above program, with the simple <b><i>Hello</i></b> class and run it.\n",
    "\n",
    "The model answer for this is called <b>Hello.py</b> and can be run with the <b>runHello.py</b> script.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello world\n"
     ]
    }
   ],
   "source": [
    "class Hello: \n",
    "    def sayHi(self): \n",
    "        print ('Hello world') \n",
    "\n",
    "h = Hello()\n",
    "h.sayHi()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='init_method'></a>\n",
    "## The __init__ method\n",
    "\n",
    "The <b><i>__init__ method</i></b> is run as soon as an object of a class is instantiated. The method can be used to do any initialization of the object you want to do. \n",
    "\n",
    "<b>NOTE:</b>  there is a double underscore both in the beginning and at the end in the name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Person: \n",
    "    def __init__(self, name): \n",
    "        self.name = name \n",
    "    def sayHi(self): \n",
    "        print ('Hello, my name is', self.name)\n",
    "\n",
    "p = Person('Stan') \n",
    "\n",
    "p.sayHi() \n",
    "# Can also be written as Person(‘Stan').sayHi()  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The <b><i>__init__ method</i></b> takes a parameter <b><i>name</i></b> (along with the usual <b><i>self</i></b>). \n",
    "\n",
    "\t def __init__(self, name): \n",
    "\n",
    "Variables that belong to an object or class are called <b><i>fields</i></b> and a new <b><i>field</i></b>, also called <b><i>name</i></b>, is created. Notice these are two different variables even though they have the same name. \n",
    "The dotted notation differentiates between them. \n",
    "\n",
    "\tself.name = name \n",
    "\n",
    "\n",
    "Most importantly, notice that the <b><i>__init__ method</i></b> is not explicitly called but the arguments in the parentheses following the class name are passed to it when creating a new instance of the class. This is the special significance of this method. \n",
    "\n",
    "\t p = Person(‘Stan') \n",
    "\n",
    "\n",
    "The <b><i>self.name</i></b> field can now be used in methods, as in the <b><i>sayHi</i></b> method. \n",
    "\n",
    "\tp.sayHi() \n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 25\n",
    "\n",
    "Write the above program, with the modified <b><i>Person</i></b> class and run it.\n",
    "\n",
    "The model answer for this is called <b>Person.py</b>.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello, my name is Abigail\n"
     ]
    }
   ],
   "source": [
    "class Person:\n",
    "    def __init__(self, name): \n",
    "            self.name = name \n",
    "\n",
    "    def sayHi(self): \n",
    "        print ('Hello, my name is', self.name) \n",
    "\n",
    "p = Person(\"Abigail\")\n",
    "p.sayHi()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='class_object_variables'></a>\n",
    "## Class and Object Variables\n",
    "\n",
    "Class variables are shared in the sense that they are accessed by all objects (instances) of that class. There is only copy of the class variable and when any one object makes a change to a class variable, the change is reflected in all the other instances as well. \n",
    "\n",
    "Object variables are owned by each individual object/instance of the class. In this case, each object has its own copy of the field i.e. they are not shared and are not related in any way to the field by the same name in a different instance of the same class. \n",
    "\n",
    "This is demonstrated in the following modification of the <b><i>Person</i></b> class:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Person: \n",
    "    #Represents a person\n",
    "    population = 0 \n",
    "    def __init__(self, name): \n",
    "    #Initializes the person's data\n",
    "        self.name = name \n",
    "        print  ('Initializing %s' % self.name)\n",
    "        # Add them to the population \n",
    "        Person.population += 1 \n",
    "\n",
    "    def sayHi(self): \n",
    "        #Greeting by the person\n",
    "        print ('Hi, my name is %s.' % self.name)\n",
    "\n",
    "    def howMany(self): \n",
    "        # Prints the current population.\n",
    "        print ('We have a population of %d here.' % Person.population)\n",
    "\n",
    "stan  = Person('Stan')\n",
    "stan.sayHi() \n",
    "stan.howMany() \n",
    "\n",
    "brian = Person('Brian')\n",
    "brian.sayHi() \n",
    "brian.howMany() \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When this is run the output will be:\n",
    "\n",
    "    (Initializing Stan) \n",
    "\tHi, my name is Stan. \n",
    "\tWe have a population of 1 here. \n",
    "\t(Initializing Brian) \n",
    "\tHi, my name is Brian.\n",
    "\tWe have a population of 2 here. \n",
    "\n",
    "The <b><i>population</i></b> variable belongs to the <b><i>Person</i></b> class and hence is a <b><i>class variable</i></b>. The <b><i>name variable</i></b> belongs to the object (it is assigned using self) and hence is an <b><i>object variable</i></b>. \n",
    "\n",
    "Therefore, refer to the population class variable as <b><i>Person.population</i></b> and not as <b><i>self.population</i></b>. \n",
    "\n",
    "Note that an object variable with the same name as a class variable will hide the class variable! Refer to the object variable name using self.\n",
    "\n",
    "Another Python variation, common to other languages, is the print formatting option shown in the classes where the information is printed:\n",
    "\n",
    "\tprint ('Initializing %s' % self.name)\n",
    "\n",
    "The %s is a placeholder for a string in the text to be printed. The string to place here is defined with a % after the enclosing print quote, in this case <b><i>% self.name</i></b>. This means that <b><i>self.name</i></b> will be printed in place of %s.\n",
    "\n",
    "It is possible to have multiple strings and by enclosing the variables in parenthesis:\n",
    "\n",
    "    first = \"Ema\"\n",
    "    second = \"Jones\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "first = \"Ema\"\n",
    "second = \"Jones\"\n",
    "\n",
    "print (\"My first name is %s and my second is %s\" %(first, second))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The %s is used to denote a string but %d can be used for a digit and %f for a float.\n",
    "\n",
    "There are options to modify these that enable text to be formatted for example to align text or fix floats to a set number of decimal places. \n",
    "\n",
    "---\n",
    "## Exercise 26\n",
    "\n",
    "Update your program to use the modified <b><i>Person</i></b> class and run it.\n",
    "\n",
    "The model answer for this is called <b>Person2.py</b>.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing Stan\n",
      "Hi, my name is Stan.\n",
      "We have a population of 1 here.\n",
      "\n",
      "Initializing Brian\n",
      "Hi, my name is Brian.\n",
      "We have a population of 2 here.\n"
     ]
    }
   ],
   "source": [
    "class Person: \n",
    "    #Represents a person\n",
    "    population = 0 \n",
    "    def __init__(self, name): \n",
    "        #Initializes the person's data\n",
    "        self.name = name \n",
    "        print  ('Initializing %s' % self.name)\n",
    "        # Add them to the population \n",
    "        Person.population += 1 \n",
    "\n",
    "    def sayHi(self): \n",
    "        #Greeting by the person\n",
    "        print ('Hi, my name is %s.' % self.name)\n",
    "\n",
    "    def howMany(self): \n",
    "        # Prints the current population.\n",
    "        print ('We have a population of %d here.' % Person.population)\n",
    "\n",
    "\n",
    "stan  = Person(\"Stan\")\n",
    "stan.sayHi() \n",
    "stan.howMany() \n",
    "print('')\n",
    "brian = Person(\"Brian\")\n",
    "brian.sayHi() \n",
    "brian.howMany() \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='inheritance'></a>\n",
    "## Inheritance\n",
    "\n",
    "As expected inheritance is supported in Python. \n",
    "\n",
    "For example, a program which has to keep track of the teachers and students in a college. They have some common characteristics such as name, age and address. They also have specific characteristics such as salary, courses and leaves for teachers and, marks and fees for students. \n",
    "\n",
    "A class called <b><i>SchoolMember</i></b> can be created and then have the teacher and student classes <b><i>inherit</i></b> from this class i.e. they will become sub-types of this type (class) and then specific characteristics can be added to these sub-types. \n",
    "\n",
    "This means that any addition/change to the functionality in SchoolMember is automatically reflected in the subtypes as well. \n",
    "\n",
    "For example, a new ID card field can be added for both teachers and students by simply adding it to the <b><i>SchoolMember</i></b> class. However, changes in the subtypes do not affect other subtypes. \n",
    "\n",
    "Another advantage is that you can refer to a teacher or student object as a <b><i>SchoolMember</i></b> object which could be useful in some situations such as counting the number of school members. \n",
    "\n",
    "This is <b><i>polymorphism</i></b> where a sub-type can be substituted in any situation where a parent type is expected i.e. the object can be treated as an instance of the parent class. \n",
    "\n",
    "The <b><i>SchoolMember</i></b> class in this situation is known as the <b><i>base class</i></b> or the <b><i>superclass</i></b>. The <b><i>Teacher</i></b> and <b><i>Student</i></b> classes are called the <b><i>derived classes</i></b> or <b><i>subclasses</i></b>. \n",
    "\n",
    "An example class to demonstrate this is shown below, including code to run it.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SchoolMember:\n",
    "    #Represents any school member\n",
    "    def __init__(self, name, age):\n",
    "        self.name = name\n",
    "        self.age = age\n",
    "        print ('Initialized SchoolMember: %s' % self.name)\n",
    "\n",
    "    def tell(self):\n",
    "        #Tell my details\n",
    "        print ('Name:\"%s\" Age:\"%s\"' % (self.name, self.age) )\n",
    "\n",
    "class Teacher(SchoolMember):\n",
    "    #Represents a teacher\n",
    "    def __init__(self, name, age, salary):\n",
    "        SchoolMember.__init__(self, name, age)\n",
    "        self.salary = salary\n",
    "        print ('Initialized Teacher: %s' % self.name)\n",
    "\n",
    "    def tell(self):\n",
    "        SchoolMember.tell(self)\n",
    "        print ('Salary: \"%d\"' % self.salary)\n",
    "\n",
    "class Student(SchoolMember):\n",
    "    #Represents a student\n",
    "    def __init__(self, name, age, marks):\n",
    "        SchoolMember.__init__(self, name, age)\n",
    "        self.marks = marks\n",
    "        print ('Initialized Student: %s' % self.name)\n",
    "\n",
    "    def tell(self):\n",
    "        SchoolMember.tell(self)\n",
    "        print ('Marks: \"%d\"' % self.marks)\n",
    "        \n",
    "\n",
    "# To create objects of the classes\n",
    "t = Teacher('Mrs. Jones', 40, 30000)\n",
    "s = Student('Stan', 22, 75)\n",
    "\n",
    "print() # prints a blank line\n",
    "\n",
    "members = [t, s]\n",
    "for member in members:\n",
    "    member.tell() # works for both Teachers and Students"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output would be:\n",
    "\n",
    "\t(Initialized SchoolMember: Mrs. Jones) \n",
    "\t(Initialized Teacher: Mrs. Jones)\n",
    "\t(Initialized SchoolMember: Stan) \n",
    "\t(Initialized Student: Stan)\n",
    "\n",
    "\tName:\"Mrs. Jones\" Age:\"40\" Salary: \"30000\" \n",
    "\tName:“Stan\" Age:\"22\" Marks: \"75\" \n",
    "\n",
    "The Teacher and Student both inherit the <b><i>tell</i></b> method from <b><i>SchoolMember</i></b> but they overload it with their own <b><i>tell</i></b> method. In these cases they also call the <b><i>SchoolMemeber tell</i></b> method but that isn’t necessary. Overloading the method in this way is polymorphism. They inherit a method but implement it differently.\n",
    "\n",
    "<b>NOTE:</b> Python does not automatically call the constructor of the superclass, you have to explicitly call it yourself. \n",
    "\n",
    "\t\n",
    "\tclass Teacher(SchoolMember):\n",
    "\t\t'''Represents a teacher.'''\n",
    "\t\tdef __init__(self, name, age, salary):\n",
    "\t\t\tSchoolMember.__init__(self, name, age)  # <---------\n",
    "\n",
    "\n",
    "Methods of the superclass can be called by prefixing the class name to the method call and then pass in the self variable along with any arguments. \n",
    "\n",
    "Can also treat instances of <b><i>Teacher</i></b> or <b><i>Student</i></b> as just instances of the <b><i>SchoolMember</i></b> when we use the tell method of the <b><i>SchoolMember</i></b> class. \n",
    "\n",
    "\n",
    "\tSchoolMember.tell(self)\n",
    "\n",
    "The classes can all be in the same file, perhaps called <b>People.py</b> and would be imported in a script with:\n",
    "\n",
    "\tfrom People import *\n",
    "\n",
    "\n",
    "## Exercise 27\n",
    " \n",
    "Write a program that implements and tests the school member classes. \n",
    "\n",
    "The model answer for this is called <b>SchoolMember.py</b>.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initialized SchoolMember: Mrs. Jones\n",
      "Initialized Teacher: Mrs. Jones\n",
      "Initialized SchoolMember: Stan\n",
      "Initialized Student: Stan\n",
      "Name:\"Mrs. Jones\" Age:\"40\"\n",
      "Salary: \"30000\"\n",
      "Name:\"Stan\" Age:\"22\"\n",
      "Marks: \"75\"\n"
     ]
    }
   ],
   "source": [
    "class SchoolMember:\n",
    "    #Represents any school member\n",
    "    def __init__(self, name, age):\n",
    "        self.name = name\n",
    "        self.age = age\n",
    "        print ('Initialized SchoolMember: %s' % self.name)\n",
    "\n",
    "    def tell(self):\n",
    "        #Tell my details\n",
    "        print ('Name:\"%s\" Age:\"%s\"' % (self.name, self.age))\n",
    "\n",
    "class Teacher(SchoolMember):\n",
    "    #Represents a teacher\n",
    "    def __init__(self, name, age, salary):\n",
    "        SchoolMember.__init__(self, name, age)\n",
    "        self.salary = salary\n",
    "        print ('Initialized Teacher: %s' % self.name)\n",
    "    def tell(self):\n",
    "        SchoolMember.tell(self)\n",
    "        print ('Salary: \"%d\"' % self.salary)\n",
    "\n",
    "class Student(SchoolMember):\n",
    "    #Represents a student\n",
    "    def __init__(self, name, age, marks):\n",
    "        SchoolMember.__init__(self, name, age)\n",
    "        self.marks = marks\n",
    "        print ('Initialized Student: %s' % self.name)\n",
    "\n",
    "    def tell(self):\n",
    "        SchoolMember.tell(self)\n",
    "        print ('Marks: \"%d\"' % self.marks)\n",
    "\n",
    "\n",
    "t = Teacher('Mrs. Jones', 40, 30000)\n",
    "s = Student('Stan', 22, 75)\n",
    "\n",
    "print # prints a blank line\n",
    "\n",
    "members = [t, s]\n",
    "for member in members:\n",
    "    member.tell() # works for both Teachers and Students\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='atom_class'></a>\n",
    "## Example Atom Class\n",
    "\n",
    "An atom is a generic term for multiple different elements. A molecule will be made of multiple atoms so would consist of multiple Atom objects. An Atom class can be defined to represent these generic characteristics\n",
    "\n",
    "Example variables would be:\n",
    "\n",
    "\tSymbol – C (carbon), N (nitrogen) etc\n",
    "\n",
    "The class could also store the atoms position in a molecule:\n",
    "\n",
    "\tx coordinate, y coordinate and z coordinate\n",
    "\n",
    "Other variables could be size, full name etc. The class could then contain methods to set and get these variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C     0.0000     1.0000     2.0000\n"
     ]
    }
   ],
   "source": [
    "class Atom:\n",
    "    def __init__(self,symbol,x,y,z):\n",
    "        self.symbol = symbol\n",
    "        self.position = (x,y,z)\n",
    "\n",
    "    def getSymbol(self):\n",
    "        return self.symbol\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '%s %10.4f %10.4f %10.4f' % (self.getSymbol(), self.position[0], self.position[1],self.position[2])\n",
    "    \n",
    "at = Atom('C',0.0,1.0,2.0)\n",
    "\n",
    "print (at)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<b><i>__repr__</i></b> is a reserved method name to represent the object. It should return a printable representation of the object and in this case it returns atom position details.\n",
    "\n",
    "<b>at = Atom('C',0.0,1.0,2.0)</b>\n",
    "\n",
    "\tCreate an atom object for carbon with the positions 0.0,1.0 and 2.0 \n",
    "\n",
    "<b>print (at)</b>\n",
    "\n",
    "\tThis will print:\n",
    "\n",
    "<b>'C'  0.0000  1.0000 2.0000</b>\n",
    "\n",
    "\t\"at\" is the atom object name and printing it will call the __repr__ method \tof the Atom class\n",
    "\n",
    "\tThis returns the string representation of the object\n",
    "\n",
    "<b>at.getSymbol()</b>\n",
    "\n",
    "\treturns the atom symbol, in this case:\n",
    "\n",
    "<b>C</b>\n",
    "\n",
    "---\n",
    "<a id='molecule_class'></a>\n",
    "## Molecule Class\n",
    "\n",
    "We can now create a <b><i>Molecule</i></b> class, which will consist of multiple <b><i>Atom</i></b> objects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Molecule name is Water  and it has 3 atoms\n",
      "O     0.0000     0.0000     0.0000\n",
      "H     0.0000     0.0000     1.0000\n",
      "H     0.0000     1.0000     0.0000\n",
      "\n"
     ]
    }
   ],
   "source": [
    "class Atom:\n",
    "    def __init__(self,symbol,x,y,z):\n",
    "        self.symbol = symbol\n",
    "        self.position = (x,y,z)\n",
    "\n",
    "    def getSymbol(self):\n",
    "        return self.symbol\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '%s %10.4f %10.4f %10.4f' % (self.getSymbol(), self.position[0], self.position[1],self.position[2])\n",
    "    \n",
    "class Molecule:\n",
    "    def __init__(self,name='Test Molecule'):\n",
    "        self.name = name\n",
    "        self.atomlist = []\n",
    "\n",
    "    def addatom(self,atom):\n",
    "        self.atomlist.append(atom)\n",
    "\n",
    "    def __repr__(self):\n",
    "        s = 'Molecule name is %s ' % self.name\n",
    "        s = s + ' and it has %d atoms\\n' % len(self.atomlist)\n",
    "        for atom in self.atomlist:\n",
    "            s = s + str(atom) + '\\n'\n",
    "        return s\n",
    "                \n",
    "\n",
    "mol = Molecule('Water')\n",
    "\n",
    "at = Atom('O',0.,0.,0.)  #  Create first atom\n",
    "mol.addatom(at)   #  Add it to the molecule object\n",
    "\n",
    "# The next atoms are added directly to the object:\n",
    "mol.addatom(Atom('H',0.,0.,1.))\n",
    "mol.addatom(Atom('H',0.,1.,0.))\n",
    "\n",
    "print (mol)  # Print the molecule object details \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ideally you would actually create the <b>Atom</b> and <b>Molecule</b> classes as separate files and then import them into your script as required:\n",
    "\n",
    "    # Creating a water molecule:\n",
    "    from Molecule import Molecule\n",
    "    from Atom import Atom \n",
    "\n",
    "    mol = Molecule('Water')\n",
    "\n",
    "    at = Atom(‘O’,0.,0.,0.)  #  Create first atom\n",
    "    mol.addatom(at)   #  Add it to the molecule object\n",
    "\n",
    "    # The next atoms are added directly to the object:\n",
    "    mol.addatom(atom(‘H’,0.,0.,1.))\n",
    "    mol.addatom(atom(‘H’,0.,1.,0.))\n",
    "\n",
    "    print (mol)  # Print the molecule object details \n",
    "    \n",
    "Files for these classes, called <b>Atom.py</b> and <b>Molecule.py</b>, are available on the exercises ansers page and can be tested with <b>runAtom.py</b> and <b>runMolecule.py</b> respectively.\n",
    "\n",
    "---\n",
    "## Exercise 28\n",
    "\n",
    "Currently the position variable in <b>Atom</b> is publicly accessible. Test the code to demonstrate the problem that this can cause. For example, write a script that changes the positions to different values. Modify the code so that the position is private and cannot be changed and test its accessibility with a script.\n",
    "\n",
    "There is no model answer for this exercise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "can't set attribute",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-d55edf47c72f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m'%s %10.4f %10.4f %10.4f'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetSymbol\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0mat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAtom\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'C'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1.0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0mat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAtom\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'C'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1.0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-d55edf47c72f>\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, symbol, x, y, z)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msymbol\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mz\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msymbol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msymbol\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mposition\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mz\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: can't set attribute"
     ]
    }
   ],
   "source": [
    "class Atom:\n",
    "    def __init__(self,symbol,x,y,z):\n",
    "        self.symbol = symbol\n",
    "        self.position = (x,y,z)\n",
    "\n",
    "    @property\n",
    "    def position(self):\n",
    "        return 0,0,0\n",
    "    \n",
    "    def getSymbol(self):\n",
    "        return self.symbol\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '%s %10.4f %10.4f %10.4f' % (self.getSymbol(), self.position[0], self.position[1],self.position[2])\n",
    "    \n",
    "at = Atom('C',0.0,1.0,2.0)\n",
    "at = Atom('C',0.0,1.0,2.0)\n",
    "\n",
    "\n",
    "print (at)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Exercise 29\n",
    "\n",
    "Download the <b>1l2y.coords</b> from the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">Exercise Answers</a> page.\n",
    "\n",
    "This is an extract from the PDB file for Trp-cage protein, which is 20 amino acids in length. The file contains the atomic coordinates for one model of the structure. \n",
    "\n",
    "Create a new class called Protein to store the protein sequence information. This class could contain a list of molecule objects but the actual design is up to you. The class should provide methods to for the following functionality: \n",
    "\n",
    "i)\tReturn the amino acid sequence of the protein. You can use the 3 letter names as provided in the coordinates file. This is the method used in the model answer but it would be better to use a dictionary to map these to the single letter codes.\n",
    "ii)\tProvide the amino acid name, atom name and atom coordinate details for any atom at a particular position. This method will take the atom position in the sequence as its argument. For example, “3” would provide all of the information for the third atom, which is a carbon in an aspartic acid.\n",
    "iii)\tProvide the amino acid name and atom coordinate details for all atoms in an amino acid at a particular position. This method will take the amino acid position in the sequence as its argument. For example, “2” would provide all of the information for the atoms in the second amino acid, which is alanine.\n",
    "\n",
    "\n",
    "For reference the full amino acid sequence is:\n",
    "\n",
    "DAYAQWLADAGWASARPPPS\n",
    "\n",
    "The format of the cords file is:\n",
    "\n",
    "ATOM      1  N   ASP A   1      -5.411  -4.929   5.995  1.00  0.00           N  \n",
    "ATOM      2  CA  ASP A   1      -6.722  -4.516   5.416  1.00  0.00           C  \n",
    "ATOM      3  C   ASP A   1      -6.568  -3.352   4.434  1.00  0.00           C  \n",
    "\n",
    "Column 2 is the atom number\n",
    "Column 4 is the amino acid and column 5 is the amino acid one letter code\n",
    "Column 6 is the aminoacid number\n",
    "Columns 7 – 9 are the X, Y and Z coordinates for the atom\n",
    "Column 12 is the atom\n",
    "\n",
    "The full PDB is also available from the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">Exercise Answers</a> page.\n",
    "\n",
    "The model answer for this is called <b>Protein.py</b> and can be tested with <b>runProtein.py</b>.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Protein:\n",
    "    def __init__(self,name='Test Molecule'):\n",
    "        self.name = name\n",
    "        self.atomlist = []\n",
    "        self.aa_dict = {}\n",
    "        self.aanum_dict = {}\n",
    "    def addatom(self,atom, aa, aanum):\n",
    "        self.atomlist.append(atom)\n",
    "        self.aa_dict[len(self.atomlist)] = aa\n",
    "        self.aanum_dict[len(self.atomlist)] = aanum\n",
    "    def addsequence(self,seq):\n",
    "        self.sequence = seq\n",
    "    def getsequence(self):\n",
    "        print (self.sequence)\n",
    "    def getatomdetails(self, atom_pos):\n",
    "        at = self.atomlist[atom_pos - 1]\n",
    "        print (self.aa_dict[atom_pos], at)\n",
    "    def getaadetails(self, aapos):\n",
    "\n",
    "        for atpos in range(0, len(self.atomlist)):\n",
    "            #print (self.aanum_dict[atpos], aapos)\n",
    "            if atpos in self.aanum_dict and int(self.aanum_dict[atpos]) == int((aapos)):\n",
    "                self.getatomdetails(atpos)\n",
    "    def __repr__(self):\n",
    "        s = 'Molecule name is %s ' % self.name\n",
    "        s = s + ' and it has %d atoms\\n' % len(self.atomlist)\n",
    "        for atom in self.atomlist:\n",
    "            s = s + str(atom) + '\\n'\n",
    "        return s\n",
    "\n",
    "class Atom:\n",
    "    def __init__(self,symbol,x,y,z):\n",
    "        self.symbol = symbol\n",
    "        self.position = (x,y,z)\n",
    "\n",
    "    def getSymbol(self):\n",
    "        return self.symbol\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '%s %10.4f %10.4f %10.4f' % (self.getSymbol(), self.position[0], self.position[1],self.position[2])\n",
    "     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Protein Sequece:\n",
      "ASN LEU TYR ILE GLN TRP LEU LYS ASP GLY GLY PRO SER SER GLY ARG PRO PRO PRO SER \n",
      "\n",
      "Atom Details:\n",
      "ASN N    -8.9010     4.1270    -0.5550\n",
      "\n",
      "AA Details\n",
      "ASN N    -8.9010     4.1270    -0.5550\n",
      "ASN C    -8.6080     3.1350    -1.6180\n",
      "ASN C    -7.1170     2.9640    -1.8970\n",
      "ASN O    -6.6340     1.8490    -1.7580\n",
      "ASN C    -9.4370     3.3960    -2.8890\n",
      "ASN C   -10.9150     3.1300    -2.6110\n",
      "ASN O   -11.2690     2.7000    -1.5240\n",
      "ASN N   -11.8060     3.4060    -3.5430\n",
      "ASN H    -8.3300     3.9570     0.2610\n",
      "ASN H    -8.7400     5.0680    -0.8890\n",
      "ASN H    -9.8770     4.0410    -0.2930\n",
      "ASN H    -8.9300     2.1620    -1.2390\n",
      "ASN H    -9.3100     4.4170    -3.1930\n",
      "ASN H    -9.1080     2.7190    -3.6790\n",
      "ASN H   -11.5720     3.7910    -4.4440\n",
      "ASN H   -12.7570     3.1830    -3.2940\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Creating a protein molecule:\n",
    "\n",
    "prot = Protein('Trp-cage')\n",
    "seq = \"\"\n",
    "\n",
    "countaa = 0\n",
    "\n",
    "\n",
    "with open (\"1l2y.coords\") as protfile:\n",
    "    for line in protfile:\n",
    "        line = line.rstrip()\n",
    "        split_line = line.split()\n",
    "        aa = split_line[3]\n",
    "        aanum = split_line[5]\n",
    "        at = Atom(split_line[11], float(split_line[6]), float(split_line[7]), float(split_line[8]))  #  Create first atom\n",
    "        prot.addatom(at, aa, aanum)   #  Add it to the molecule object\n",
    "\n",
    "        if(int(aanum) > countaa):\n",
    "            seq += aa + \" \"\n",
    "            countaa+=1\n",
    "\n",
    "prot.addsequence(seq)\n",
    "\n",
    "# print (prot)  # Print the molecule object details \n",
    "\n",
    "print (\"\\nProtein Sequece:\")\n",
    "prot.getsequence()\n",
    "\n",
    "print (\"\\nAtom Details:\")\n",
    "prot.getatomdetails(1)\n",
    "\n",
    "print(\"\\nAA Details\")\n",
    "prot.getaadetails(1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='sequence_files'></a>\n",
    "# Handling Sequence Files\n",
    "\n",
    "An example problem may be the need to write a program to parse a multi fasta sequence file and extract a specific sequence based on accession ID. For example, extract P01322 from a file with the following format:\n",
    "\n",
    "\\>sp|Q91XI3|INS_SPETR Insulin OS=Spermophilus tridecemlineatus GN=INS PE=3 SV=1\n",
    "MALWTRLLPLLALLALLGPDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRREVEEQQGGQVELGGGPGAGLPQPLALEMALQKRGIVEQCCTSICSLYQLENYCN<br>\n",
    "\\>sp|P01313|INS_CRILO Insulin OS=Cricetulus longicaudatus GN=INS PE=1 SV=2\n",
    "MTLWMRLLPLLTLLVLWEPNPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRRGVEDPQVAQLELGGGPGADDLQTLALEVAQQKRGIVDQCCTSICSLYQLENYCN<br>\n",
    "\\>sp|P01322|INS1_RAT Insulin-1 OS=Rattus norvegicus GN=Ins1 PE=1 SV=1\n",
    "MALWMRFLPLLALLVLWEPKPAQAFVKQHLCGPHLVEALYLVCGERGFFYTPKSRREVEDPQVPQLELGGGPEAGDLQTLALEVARQKRGIVDQCCTSICSLYQLENYCN<br>\n",
    "\\>sp|P01315|INS_PIG Insulin OS=Sus scrofa GN=INS PE=1 SV=2\n",
    "MALWTRLLPLLALLALWAPAPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREAENPQAGAVELGGGLGGLQALALEGPPQKRGIVEQCCTSICSLYQLENYCN<br>\n",
    "......\n",
    "\n",
    "A procedural example would be:\n",
    "\n",
    "    with open('seq.fasta') as in_file:\n",
    "        match = 0\n",
    "            for line in in_file:\n",
    "                line = line.rstrip()  # Remove newline \n",
    "                if line.startswith('>'):\n",
    "                    # A description line so search for required ID\n",
    "                if line.find('P01322') != -1:\n",
    "                    # It is the required sequence so set flag to 1\n",
    "                \tmatch = 1\n",
    "            \telse:\n",
    "                \t# Not the required sequence so set flag to 0\n",
    "                \tmatch = 0\n",
    "        \t\tif match:\n",
    "            \t\t# In the correct sequence so print it out\n",
    "            \t\tprint(line)\n",
    "\n",
    "\n",
    "Not the most elegant solution but it fulfils the requirement, even if it is not immediately clear how it is working. The code is also limited and difficult to add further functionality\n",
    "\n",
    "It may be required to add functionality to return the length of the sequence, handle nucleotide as well as protein sequences including options to translate them former and extracting multiple sequences based on a list of IDs.\n",
    "\n",
    "All of these are achievable but it is not immediately obvious how to add them to the code. They will also make the code more complex and difficult to understand.\n",
    "\n",
    "OO provides an easier mechanism to add additional functionality and also ensures the code is easier to understand. Object oriented programming groups data and functionality together and provides the ability to abstract away some of the complexities of the processing logic and to encapsulate the data\n",
    "\n",
    "Two classes will provide this functionality:\n",
    "\n",
    "<b>FastaRecord</b> - Store and provide the sequence information \n",
    "<b>FastaParser</b> - Parse the fasta record to create the FastaRecord objects \n",
    "\n",
    "To test the following code you'll need the <b>multi_sequences.fasta</b> file from the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">exercise anwsers</a> page. You need to store this file in the same directory as the Jupyter Notebook files.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class FastaRecord():\n",
    "    # Class representing a FASTA record\n",
    "\n",
    "    def __init__(self, description_line):\n",
    "        # Initialise an instance of the FastaRecord clas\n",
    "        self.description = description_line.strip()\n",
    "        self.sequences = []\n",
    "\n",
    "    def add_sequence_line(self, sequence_line):\n",
    "\n",
    "        # Add a sequence line to the FastaRecord instance.\n",
    "        # This function can be called more than once.\n",
    "\n",
    "        self.sequences.append( sequence_line.strip() )\n",
    "\n",
    "    def __repr__(self):\n",
    "        # Representation of the FastaRecord instance\n",
    "        lines = [self.description,]\n",
    "        lines.extend(self.sequences)\n",
    "        return '\\n'.join(lines)\n",
    "\n",
    "\n",
    "class FastaParser():\n",
    "    # Class for parsing FASTA files\n",
    "\n",
    "    def __init__(self, fpath):\n",
    "        # Initialise an instance of the FastaParser\n",
    "        self.fpath = fpath\n",
    "\n",
    "    def __iter__(self):\n",
    "        # Yield FastaRecord instances\n",
    "        fasta_record = None\n",
    "        with open(self.fpath, 'r') as fh:\n",
    "            for line in fh:\n",
    "                if line.startswith('>'):\n",
    "                    if fasta_record:\n",
    "                        yield fasta_record\n",
    "                    fasta_record = FastaRecord(line)\n",
    "                else:\n",
    "                    fasta_record.add_sequence_line(line)\n",
    "        yield fasta_record\n",
    "\n",
    "for fasta_record in FastaParser('multi_sequences.fasta'):\n",
    "    print(fasta_record.description)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Both classes can be in the same file and then imported into the test script. Assuming they are in a file called <b>Fasta.py</b>:\n",
    "\n",
    "    from Fasta import FastaParser\n",
    "\n",
    "    for fasta_record in FastaParser('multi_sequences.fasta'):\n",
    "       \t\tprint(fasta_record.description)\n",
    "\n",
    "\n",
    "The <b>Fasta.py</b> file and a testing script called <b>runFasta.py</b> are available from the the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">exercise anwsers</a> page.\n",
    "\n",
    "---\n",
    "<a id='iter'></a>\n",
    "## __iter__\n",
    "\n",
    "<b>def __iter__(self):</b>\n",
    "\n",
    "A special class method that provides an iteration \n",
    "\n",
    "For example:\n",
    "\n",
    "\tfor fasta_record in FastaParser('example.fasta'):\n",
    "        print(fasta_record.description)\n",
    "\n",
    "\n",
    "This creates a <b>FastaParser</b> object with the fasta sequence file <b><i>example.fasta</i></b>. As the for loop is an iteration being called on FastaParser it automatically calls the <b>__iter__</b> method. This is the same as print automatically calling the <b>__repr__ method</b>.\n",
    "\n",
    "---\n",
    "<a id='magic_methods'></a>\n",
    "## Magic Methods\n",
    "\n",
    "<b>__repr__</b> and <b>__iter__</b> are examples of magic methods. The <i>magic</i> aspect of these methods is that you don’t need to call them explicitly. A for loop calls <b>__iter__</b> and print calls <b>__repr__</b>.\n",
    "\n",
    "There are many magic methods that can be used and they are not well documented but they are explained at:\n",
    "\n",
    "\thttp://www.rafekettler.com/magicmethods.html\n",
    "\n",
    "Magic methods are used all the time in Python but as they are called silently they are not noticed:\n",
    "\n",
    "\t+ \t\tobject.__add__(self, other)\n",
    "\t*= \t\tobject.__imul__(self, other) \n",
    "\tint() \t\tobject.__int__(self) \n",
    "\t> \t\tobject.__gt__(self, other) \n",
    "\tetc\n",
    "    \n",
    "---    \n",
    "<a id='iterables'></a>\n",
    "## Iterables\n",
    "\n",
    "To explain <b><i>yield</i></b> first need to understand <b><i>iterables</i></b> and <b><i>generators</i></b>. When you create a <b><i>list</i></b> you can read it one element at a time in a <b><i>for loop</i></b> (iteration) and anything that can be read in a <b><i>for loop</i></b> is an <b><i>iterable</i></b> (<b><i>list, string, file</i></b> etc). They can be read as many times as you like as they are stored in memory. However, <b><i>iterables</i></b> are not efficient if you need to store a large amount of data and particular if you only want to iterate over them once.\n",
    "\n",
    "---\n",
    "<a id='generators'></a>\n",
    "## Generators\n",
    "\n",
    "<b><i>Generators</i></b> are <b><i>iterators</i></b>, but you can only iterate over them once\n",
    "\n",
    "They do not store all the values in memory and instead generate them on the fly:\n",
    "\n",
    " \tmygenerator = (x*x for x in range(3))\n",
    "\tfor i in mygenerator:\n",
    "\t\tprint(i)\n",
    "\n",
    "\n",
    "Cannot perform <b><i>\"for i in mygenerator\"</i></b> a second time as <b><i>generators</i></b> can only be used once. Each value is calculated once and then forgotten.\n",
    "\n",
    "---\n",
    "<a id='yield'></a>\n",
    "## Yield\n",
    "\n",
    "<b><i>Yield</i></b> is a special return keyword that returns a <b><i>generator</i></b>.\n",
    "\n",
    "In the <b>FastaParser</b> class the <b><i>yield</i></b> keyword is required as the magic method <b><i>__iter__</i></b> is an <b><i>iteration</i></b>. The method returns a <b><i>generator</i></b> and so effectively each fasta record is returned individually and they are not stored:\n",
    "    \n",
    "    for line in in_file :\n",
    "        if line.startswith('>'):\n",
    "        \tif fasta_record:\n",
    "           \t\tyield fasta_record\n",
    "           \t\tfasta_record = FastaRecord(line)\n",
    "          \telse:\n",
    "          \t\tfasta_record.add_sequence_line(line)\n",
    "        yield fasta_record\n",
    "    \n",
    "---   \n",
    "## Expanding the Fasta Classes\n",
    "\n",
    "The FastaRecord and FastaParser classes could now be extended to include additional methods\n",
    "\n",
    "<b>FastaRecord:</b>\n",
    "\n",
    "Returns a fasta sequence with a specified ID\n",
    "Returns the length of a sequence with a specified ID\n",
    "Return a translation of a specified sequence\n",
    "\n",
    "<b>FastaParser:</b>\n",
    "\n",
    "Returns the description lines of all the sequences\n",
    "Returns the IDs of all the sequences\n",
    "\n",
    "---\n",
    "## Benefits of the Object Oriented Version\n",
    "\n",
    "Writing an OO version of the fasta parsing program does take longer and require more code but there are distinct advantages.\n",
    "\n",
    "It is a lot easier to understand as the functionality is grouped.\n",
    "\n",
    "It is extendible – additional functions can be added with relative ease \twhereas with the procedural version it is not so clear where to start.\n",
    "\n",
    "It is easy to use – once written the class methods are just called.\n",
    "\n",
    "---\n",
    "<a id='oo_exercises'></a>\n",
    "## Exercises\n",
    "\n",
    "Note that the model answers for these exercises are the simplest solutions but could be improved. For example, many of the variables could be made private with associated getter and setter methods.\n",
    "\n",
    "---\n",
    "## Exercise 30\n",
    "\n",
    "\n",
    "Currently the <b>FastaParser</b> class reads the fasta file each time it is used, so the <b>FastaRecord</b> objects are transient and not stored. Modify the code so that an instance of the <b>FastaParser</b> class creates and stores <b>FastaRecord</b> objects, which can then be retrieved. Write a script that tests the modified class.\n",
    "\n",
    "The model answer for this is called <b>Fasta2.py</b> and can be run with the <b>runFasta2.py</b> script.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "class FastaRecord(object):\n",
    "    # Class representing a FASTA record\n",
    "\n",
    "    def __init__(self, description_line):\n",
    "        # Initialise an instance of the FastaRecord class\n",
    "        self.sequence = description_line\n",
    "        # Remove > and whitespace on ends\n",
    "        self.description = description_line[1:]\n",
    "        self.description = self.description.strip()\n",
    "        self.seqid = (self.description.split())[0]\n",
    "        self.seq_length = 0\n",
    "\n",
    "    def add_sequence_line(self, sequence_line):\n",
    "\n",
    "        # Add a sequence line to the FastaRecord instance.\n",
    "        # This function can be called more than once.\n",
    "\n",
    "        self.seq_length += len(sequence_line.strip())\n",
    "        self.sequence += sequence_line\n",
    "\n",
    "    def __repr__(self):\n",
    "        # Representation of the FastaRecord instance\n",
    "        return self.sequence\n",
    "\n",
    "\n",
    "class FastaParser(object):\n",
    "    # Class for parsing FASTA files\n",
    "    def __init__(self, fpath):\n",
    "    # Initialise an instance of the FastaParser\n",
    "        self.fpath = fpath\n",
    "        self.record_list = []\n",
    "\n",
    "        fasta_record = None\n",
    "        with open(self.fpath, 'r') as fh:\n",
    "            for line in fh:\n",
    "                if line.startswith('>'):\n",
    "                    if fasta_record:\n",
    "                        self.record_list.append(fasta_record)\n",
    "                    fasta_record = FastaRecord(line)\n",
    "                else:\n",
    "                    fasta_record.add_sequence_line(line)\n",
    "            self.record_list.append(fasta_record)\n",
    "\n",
    "\n",
    "    def __repr__(self):\n",
    "        seqs = \"\"\n",
    "        for record in self.record_list:\n",
    "            seqs += str(record) + \"\\n\"\n",
    "        return seqs\n",
    "\n",
    "    def getRecords(self):\n",
    "        return self.record_list\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now the list of fasta records\n",
      ">sp|P01308|INS_HUMAN Insulin OS=Homo sapiens \n",
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "LQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|Q6YK33|INS_GORGO Insulin OS=Gorilla gorilla gorilla \n",
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "LQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|Q8HXV2|INS_PONPY Insulin OS=Pongo pygmaeus \n",
      "MALWMRLLPLLALLALWGPDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "LQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P30410|INS_PANTR Insulin OS=Pan troglodytes \n",
      "MALWMRLLPLLVLLALWGPDPASAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "LQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P30406|INS_MACFA Insulin OS=Macaca fascicularis \n",
      "MALWMRLLPLLALLALWGPDPAPAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "PQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P30407|INS_CHLAE Insulin OS=Chlorocebus aethiops \n",
      "MALWMRLLPLLALLALWGPDPVPAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAED\n",
      "PQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P67972|INS_AOTTR Insulin OS=Aotus trivirgatus \n",
      "MALWMHLLPLLALLALWGPEPAPAFVNQHLCGPHLVEALYLVCGERGFFYAPKTRREAED\n",
      "LQVGQVELGGGSITGSLPPLEGPMQKRGVVDQCCTSICSLYQLQNYCN\n",
      "\n",
      ">sp|P01321|INS_CANFA Insulin OS=Canis familiaris \n",
      "MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVED\n",
      "LQVRDVELAGAPGEGGLQPLALEGALQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|Q91XI3|INS_SPETR Insulin OS=Spermophilus tridecemlineatus \n",
      "MALWTRLLPLLALLALLGPDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRREVEE\n",
      "QQGGQVELGGGPGAGLPQPLALEMALQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P01313|INS_CRILO Insulin OS=Cricetulus longicaudatus \n",
      "MTLWMRLLPLLTLLVLWEPNPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRRGVED\n",
      "PQVAQLELGGGPGADDLQTLALEVAQQKRGIVDQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P01322|INS1_RAT Insulin-1 OS=Rattus norvegicus \n",
      "MALWMRFLPLLALLVLWEPKPAQAFVKQHLCGPHLVEALYLVCGERGFFYTPKSRREVED\n",
      "PQVPQLELGGGPEAGDLQTLALEVARQKRGIVDQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P01315|INS_PIG Insulin OS=Sus scrofa \n",
      "MALWTRLLPLLALLALWAPAPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREAEN\n",
      "PQAGAVELGGGLGGLQALALEGPPQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|P01311|INS_RABIT Insulin OS=Oryctolagus cuniculus \n",
      "MASLAALLPLLALLVLCRLDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRREVEE\n",
      "LQVGQAELGGGPGAGGLQPSALELALQKRGIVEQCCTSICSLYQLENYCN\n",
      "\n",
      ">sp|Q62587|INS_PSAOB Insulin OS=Psammomys obesus \n",
      "MALWMRLLPLLAFLILWEPSPAHAFVNQHLCGSHLVEALYLVCGERGFFYTPKFRRGVDD\n",
      "PQMPQLELGGSPGAGDLRALALEVARQKRGIVEQCCTGICSLYQLENYCN\n",
      "\n",
      ">sp|P06306|INS_FELCA Insulin OS=Felis catus \n",
      "MAPWTRLLPLLALLSLWIPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREAED\n",
      "LQGKDAELGEAPGAGGLQPSALEAPLQKRGIVEQCCASVCSLYQLEHYCN\n",
      "\n",
      ">sp|P01317|INS_BOVIN Insulin OS=Bos taurus \n",
      "MALWTRLRPLLALLALWPPPPARAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEG\n",
      "PQVGALELAGGPGAGGLEGPPQKRGIVEQCCASVCSLYQLENYCN\n",
      "\n",
      ">sp|P01318|INS_SHEEP Insulin OS=Ovis aries \n",
      "MALWTRLVPLLALLALWAPAPAHAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEG\n",
      "PQVGALELAGGPGAGGLEGPPQKRGIVEQCCAGVCSLYQLENYCN\n",
      "\n",
      ">sp|P17715|INS_OCTDE Insulin OS=Octodon degus \n",
      "MAPWMHLLTVLALLALWGPNSVQAYSSQHLCGSNLVEALYMTCGRSGFYRPHDRRELEDL\n",
      "QVEQAELGLEAGGLQPSALEMILQKRGIVDQCCNNICTFNQLQNYCNVP\n",
      "\n",
      ">sp|P01329|INS_CAVPO Insulin OS=Cavia porcellus \n",
      "MALWMHLLTVLALLALWGPNTGQAFVSRHLCGSNLVETLYSVCQDDGFFYIPKDRRELED\n",
      "PQVEQTELGMGLGAGGLQPLALEMALQKRGIVDQCCTGTCTRHQLQSYCN\n",
      "\n",
      ">sp|P67970|INS_CHICK Insulin OS=Gallus gallus \n",
      "MALWIRSLPLLALLVFSGPGTSYAAANQHLCGSHLVEALYLVCGERGFFYSPKARRDVEQ\n",
      "PLVSSPLRGEAGVLPFQQEEYEKVKRGIVEQCCHNTCSLYQLENYCN\n",
      "\n",
      ">sp|P12706|INS1_XENLA Insulin-1 OS=Xenopus laevis \n",
      "MALWMQCLPLVLVLFFSTPNTEALVNQHLCGSHLVEALYLVCGDRGFFYYPKVKRDMEQA\n",
      "LVSGPQDNELDGMQLQPQEYQKMKRGIVEQCCHSTCSLFQLESYCN\n",
      "\n",
      ">sp|Q98TA8|INS_PANBU Insulin OS=Pantodon buchholzi \n",
      "MALWLQAFTLLVLLVLSSPGAQSASSQHLCGSHLVDALYMVCGEKGFFYQPKTKRDVDPL\n",
      "LGFLSPKSAQENEADEYPYKDQGDLKVKRGIVEQCCHHPCNIFDLQNYCN\n",
      "\n",
      ">sp|P51463|INS_SELRF Insulin (Fragment) OS=Selasphorus rufus \n",
      "IQSLPLLALLALSGPGTSHAAVNQHLCGSHLVEALYLVCGERGFFYSPKARRDAEHPLVN\n",
      "GPLHGEVGDLPFQQEEFEKVKRGIVEQCCHNTCSLYQLENYCN\n",
      "\n",
      ">tr|R0L6M4|R0L6M4_ANAPL Insulin (Fragment) OS=Anas platyrhynchos \n",
      "MALWIRSLPLLALLALSGPGISHAAANQHLCGSHLVEALYLVCGERGFFYSPKTRRDVEQ\n",
      "PLVNGPLHGEVGELPFQHEEYQKVKRGIVEQCCENPCSLYQLENYCN\n"
     ]
    }
   ],
   "source": [
    "parser = FastaParser('multi_sequences.fasta')\n",
    "# print (parser)\n",
    "\n",
    "print (\"Now the list of fasta records\")\n",
    "\n",
    "for fasta_record in parser.getRecords():\n",
    "\tprint(fasta_record)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Exercise 31\n",
    "\n",
    "Modify the code to include the following functionality: \n",
    "\t\n",
    "Returns a fasta sequence with a specified ID.\n",
    "Returns the description lines of all the sequences as a list.\n",
    "Returns the IDs of all the sequences as a list.\n",
    "Returns the length of a sequence with a specified ID.\n",
    "\n",
    "To incorporate this functionality separate methods should be added to the <b>FastaRecord</b> class. Write a script that tests the modified class.\n",
    "\n",
    "The model answer for this is called <b>Fasta3.py</b> and can be run with the <b>runFasta3.py</b> script.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "class FastaRecord(object):\n",
    "    # Class representing a FASTA record\n",
    "\n",
    "\tdef __init__(self, description_line):\n",
    "\t\t# Initialise an instance of the FastaRecord class\n",
    "\t\tself.sequence = description_line\n",
    "\t\t# Remove > and whitespace on ends\n",
    "\t\tself.description = description_line[1:]\n",
    "\t\tself.description = self.description.strip()\n",
    "\t\tself.seqid = (self.description.split())[0]\n",
    "\t\tself.seq_length = 0\n",
    "\n",
    "\tdef add_sequence_line(self, sequence_line):\n",
    "        \n",
    "\t\t# Add a sequence line to the FastaRecord instance.\n",
    "\t\t# This function can be called more than once.\n",
    "        \n",
    "\t\tself.seq_length += len(sequence_line.strip())\n",
    "\t\tself.sequence += sequence_line\n",
    "\n",
    "\tdef matchesID(self, searchid):\n",
    "\t\t# Return description line by ID\n",
    "\t\tif self.seqid == searchid:\n",
    "\t\t\treturn True\n",
    "\n",
    "\tdef getDescription(self):\n",
    "\t\t# Return description line\n",
    "\t\treturn self.description \n",
    "\n",
    "\tdef getSeqLength(self):\n",
    "\t\t# Return seq_length\n",
    "\t\treturn self.seq_length\n",
    "    \n",
    "\tdef __repr__(self):\n",
    "\t\t# Representation of the FastaRecord instance\n",
    "\t\treturn self.sequence\n",
    "\n",
    "\n",
    "class FastaParser(object):\n",
    "   \t# Class for parsing FASTA files\n",
    "\tdef __init__(self, fpath):\n",
    "\t# Initialise an instance of the FastaParser\n",
    "\t\tself.fpath = fpath\n",
    "\t\tself.record_list = []\n",
    "\n",
    "\t\tfasta_record = None\n",
    "\t\twith open(self.fpath, 'r') as fh:\n",
    "\t\t\tfor line in fh:\n",
    "\t\t\t\tif line.startswith('>'):\n",
    "\t\t\t\t\tif fasta_record:\n",
    "\t\t\t\t\t\tself.record_list.append(fasta_record)\n",
    "\t\t\t\t\tfasta_record = FastaRecord(line)\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tfasta_record.add_sequence_line(line)\n",
    "\t\t\tself.record_list.append(fasta_record)\n",
    "\n",
    "\t\n",
    "\tdef __repr__(self):\n",
    "\t\tseqs = \"\"\n",
    "\t\tfor record in self.record_list:\n",
    "\t\t\tseqs += str(record) + \"\\n\"\n",
    "\t\treturn seqs\n",
    "\n",
    "\tdef getRecords(self):\n",
    "\t\treturn self.record_list\n",
    "\n",
    "\tdef getRecordByID(self, seqid):\n",
    "\t\tfor record in self.record_list:\n",
    "\t\t\tif record.matchesID(seqid):\n",
    "\t\t\t\treturn record\n",
    "\n",
    "\tdef getSeqLength(self, seqid):\n",
    "\t\tfor record in self.record_list:\n",
    "\t\t\tif record.matchesID(seqid):\n",
    "\t\t\t\treturn record.getSeqLength()\n",
    "\n",
    "\tdef getIDs(self):\n",
    "\t\tids = []\n",
    "\t\tfor record in self.record_list:\n",
    "\t\t\tids.append(record.seqid)\n",
    "\t\treturn ids\n",
    "\n",
    "\tdef getDescriptions(self):\n",
    "\t\tdescriptions = []\n",
    "\t\tfor record in self.record_list:\n",
    "\t\t\tdescriptions.append(record.description)\n",
    "\t\treturn descriptions\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now the list of fasta records\n",
      "Record by ID\n",
      "None\n",
      "['sp|P01308|INS_HUMAN Insulin OS=Homo sapiens', 'sp|Q6YK33|INS_GORGO Insulin OS=Gorilla gorilla gorilla', 'sp|Q8HXV2|INS_PONPY Insulin OS=Pongo pygmaeus', 'sp|P30410|INS_PANTR Insulin OS=Pan troglodytes', 'sp|P30406|INS_MACFA Insulin OS=Macaca fascicularis', 'sp|P30407|INS_CHLAE Insulin OS=Chlorocebus aethiops', 'sp|P67972|INS_AOTTR Insulin OS=Aotus trivirgatus', 'sp|P01321|INS_CANFA Insulin OS=Canis familiaris', 'sp|Q91XI3|INS_SPETR Insulin OS=Spermophilus tridecemlineatus', 'sp|P01313|INS_CRILO Insulin OS=Cricetulus longicaudatus', 'sp|P01322|INS1_RAT Insulin-1 OS=Rattus norvegicus', 'sp|P01315|INS_PIG Insulin OS=Sus scrofa', 'sp|P01311|INS_RABIT Insulin OS=Oryctolagus cuniculus', 'sp|Q62587|INS_PSAOB Insulin OS=Psammomys obesus', 'sp|P06306|INS_FELCA Insulin OS=Felis catus', 'sp|P01317|INS_BOVIN Insulin OS=Bos taurus', 'sp|P01318|INS_SHEEP Insulin OS=Ovis aries', 'sp|P17715|INS_OCTDE Insulin OS=Octodon degus', 'sp|P01329|INS_CAVPO Insulin OS=Cavia porcellus', 'sp|P67970|INS_CHICK Insulin OS=Gallus gallus', 'sp|P12706|INS1_XENLA Insulin-1 OS=Xenopus laevis', 'sp|Q98TA8|INS_PANBU Insulin OS=Pantodon buchholzi', 'sp|P51463|INS_SELRF Insulin (Fragment) OS=Selasphorus rufus', 'tr|R0L6M4|R0L6M4_ANAPL Insulin (Fragment) OS=Anas platyrhynchos']\n",
      "['sp|P01308|INS_HUMAN', 'sp|Q6YK33|INS_GORGO', 'sp|Q8HXV2|INS_PONPY', 'sp|P30410|INS_PANTR', 'sp|P30406|INS_MACFA', 'sp|P30407|INS_CHLAE', 'sp|P67972|INS_AOTTR', 'sp|P01321|INS_CANFA', 'sp|Q91XI3|INS_SPETR', 'sp|P01313|INS_CRILO', 'sp|P01322|INS1_RAT', 'sp|P01315|INS_PIG', 'sp|P01311|INS_RABIT', 'sp|Q62587|INS_PSAOB', 'sp|P06306|INS_FELCA', 'sp|P01317|INS_BOVIN', 'sp|P01318|INS_SHEEP', 'sp|P17715|INS_OCTDE', 'sp|P01329|INS_CAVPO', 'sp|P67970|INS_CHICK', 'sp|P12706|INS1_XENLA', 'sp|Q98TA8|INS_PANBU', 'sp|P51463|INS_SELRF', 'tr|R0L6M4|R0L6M4_ANAPL']\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "\n",
    "parser = FastaParser('multi_sequences.fasta')\n",
    "# print (parser)\n",
    "\n",
    "#for fasta_record in parser:\n",
    "#\tprint(fasta_record.description)\n",
    "    \t#if fasta_record.matches('Q9Y233'):\n",
    "        #\tprint(fasta_record)\n",
    "\n",
    "print (\"Now the list of fasta records\")\n",
    "\n",
    "# for fasta_record in parser.getRecords():\n",
    "# \tprint(fasta_record)\n",
    "\n",
    "\n",
    "# Get sequence by ID\n",
    "print (\"Record by ID\")\n",
    "\n",
    "print(parser.getRecordByID(\"DS572233\"))\n",
    "\n",
    "# Get the description lines\n",
    "\n",
    "descriptions = parser.getDescriptions()\n",
    "print(descriptions)\n",
    "\n",
    "# Get the IDs\n",
    "\n",
    "ids = parser.getIDs()\n",
    "print(ids)\n",
    "\n",
    "# Get sequence length by ID\n",
    "\n",
    "seqlen = parser.getSeqLength(\"DS572233\")\n",
    "print(seqlen)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Exercise 32\n",
    "\n",
    "The classes so far are designed to handle multiple fasta sequences but sometimes a requirement is just to manipulate a single sequence. Below is a simple class to store a nucleotide sequence, which can be modified to provide additional functionality. This class will just store the sequence itself and not the description line.\n",
    "\n",
    "\tclass DNA: \n",
    "     \tdef __init__(self, s): \n",
    "        \t# Create DNA instance initialized to string s\n",
    "        \tself.seq = s \n",
    "      \n",
    "Add methods to this class to perform the following functions:\n",
    "\n",
    "Reverse the sequence.\n",
    "Complement the sequence.\n",
    "Reverse complement the sequence.\n",
    "Provide the %GC content of the sequence.\n",
    "Provide a list of codons in the sequence.\n",
    "You could also additionally add a method to translate the sequence. You should have the code for this from a previous exercise.\n",
    "\n",
    "Write a script that tests the DNA class. You can use the <b>sequence.fasta</b> file available from the the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">exercise anwsers</a> page to test the class. You will need to extract just the sequence from this file and use that to create an object instance of the DNA class.\n",
    "\n",
    "The model answer for this is called <b>DNA.py</b> and can be run with the runDNA.py script. It does not include the final option to translate the sequence, the code for which has already been provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DNA: \n",
    " \n",
    "\tdef __init__(self, s): \n",
    "\t\t# Create DNA instance initialized to by the sequence s\n",
    "\t\tself.seq = s \n",
    "\t\tself.comp = {'a':'t', 't':'a', 'g':'c', 'c':'g'}\n",
    "\t\n",
    "\tdef reverse(self): \n",
    "        \t# Return the reversed sequence\n",
    "\t\tself.revseq = self.seq[::-1]\n",
    "\t\treturn self.revseq \n",
    "\n",
    "\tdef complement(self): \n",
    "\t\t# Return the complementary sequence\n",
    "\t\tcompseq = \"\"\n",
    "\t\tfor nuc in self.seq:\n",
    "\t\t\tcompseq += self.comp[nuc] \n",
    "\t\treturn compseq \n",
    "     \n",
    "\tdef reversecomplement(self): \n",
    "\t\t# Return the reverse complement of the sequence\n",
    "\t\tcompseq = \"\"\n",
    "\t\tfor nuc in self.seq:\n",
    "\t\t\tcompseq += self.comp[nuc]\n",
    "\t\treturn compseq[::-1] \n",
    "     \n",
    "\tdef gc(self): \n",
    "\t\t# Return the percentage of sequence composed of G and C\n",
    "\t\t# Note the limitation of these next two methods are that they are case sensitive\n",
    "\t\t# They need to be modified to work with a sequence in upper case\n",
    "\t\tgc = self.seq.count('g') + self.seq.count('c') \n",
    "\t\treturn gc * 100.0 / len(self.seq) \n",
    " \n",
    "\tdef codons(self): \n",
    "\t\t# Return a list of codons for the sequence\n",
    "\t\tcodons = []\n",
    "\t\tfor count in range(0, len(self.seq), 3):  \n",
    "\t\t\tcodons.append(self.seq[count:count+3])\n",
    "\t\t \n",
    "\t\treturn codons \n",
    "\n",
    "\t# Return the printable representation of the object, the sequence\n",
    "\tdef __repr__(self):\n",
    "\t\treturn self.seq\n",
    "  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['atg', 'ttc', 'aaa', 'tcc']"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence = 'atgttcaaatcc'\n",
    "dna = DNA(sequence)\n",
    "rev = dna.reversecomplement\n",
    "codon = dna.codons\n",
    "codon()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
