{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Tutorial - Part 2\n",
    "\n",
    "---\n",
    "<a id='translate_seq'></a>\n",
    "# Example Program\n",
    "\n",
    "# Translating a Nucleotide Sequence\n",
    "\n",
    "---\n",
    "\n",
    "A common task in bioinformatics is using a scripting language, such as Python, to also parse the output from an analysis program, such as BLAST, and produce a file of edited data.\n",
    "\n",
    "Scripting languages can be used to rapidly write code to read files, manipulate the text and produce an output.\n",
    "\n",
    "An example might be to read a fasta file of nucleotide sequence and produce the possible amino acid sequence. In other words, translate the sequence.\n",
    "\n",
    "This will demonstrate the use of file handling, control structures, string (list) manipulation and dictionaries.\n",
    "\n",
    "---\n",
    "\n",
    "## Input Files\n",
    "\n",
    "---\n",
    "\n",
    "For the example we will have 2 input files.\n",
    "\n",
    "The fasta sequence:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;>BF246290<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ggcgtcgtagtctcctgcagcgtctggggtttccgttgcagtcctcggaaccaggacctc<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ggcgtggcctagcgagttatggcgacgaaggccgtgtgcgtgctgaagggcgacgg<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;cccagtgcagggctatcatcaattcgagcagaaggaaagtaatggcaccagtgaag<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;…<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;cacagatggtgtgccgatgtgtctatggaacgattctgtgatctcactctcaggagacca<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;tggcatcatgtggccgcacaactgtggtccatgaaaaagcaagatgactgtgggcca<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ggg</b><br>\n",
    "\n",
    "NOTE: This is not one complete sequence, there are line breaks\n",
    "\n",
    "A file of codon translations:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ttt<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;F<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ttc<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;F<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;tta<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;L<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ttg<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;L<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;etc</b><br>\n",
    "\n",
    "---\n",
    "\n",
    "## Storing the Sequence\n",
    "\n",
    "---\n",
    "\n",
    "The first step is to read the fasta sequence from file into a string:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;with open('sequence.txt') as in_file:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Open the file<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq_list = in_file.readlines()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Read the file to a list<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq_name = seq_list.pop(0)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Remove the sequence name<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq = seq_list.pop(0)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Initialise the sequence<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq = seq.rstrip()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Remove the newline character<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq = seq.lower()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Lower case the sequence<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for line in seq_list:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;#  Append the rest of the sequence to seq<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq += line.rstrip().lower()</b><br>\n",
    "\n",
    "NOTE: Methods can be appended to each other - seq += line.rstrip().lower()\n",
    "\n",
    "---\n",
    "\n",
    "## Building the Dictionary\n",
    "\n",
    "---\n",
    "\n",
    "The next step is to build the dictionary using the codons file:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;with open('codons.txt') as in_file:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Open the file<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codons_list = in_file.readlines()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Read the file to a list<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codons = {}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Initialise the dictionary<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for count in range(0, len(codons_list), 2):&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Work through the list<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;key = codons_list[count].rstrip()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Get the codon<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;key = key.lower()<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;value = codons_list[count+1].rstrip()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Get the aa, on next line<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;value = value.lower()<br>   \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codons[key] = value&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Add to dictionary</b><br><br>\n",
    "\n",
    "---\n",
    "\n",
    "## Translating the Sequence\n",
    "\n",
    "---\n",
    "\n",
    "Now we have the sequence and codon translations we create the amino acid sequence:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for count in range(0, len(seq), 3):&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Work through the list<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codon = seq[count:count+3]&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Get 3 nucleotides<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aa = codons[codon]&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# Get the associated aa<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;print (aa, end=\" \")<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# In 2.7 this would be:<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# print aa,</b><br><br>\n",
    "\n",
    "---\n",
    "\n",
    "## Complete Program\n",
    "\n",
    "---\n",
    "\n",
    "Putting it all together, with a few short cuts:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;with open('sequence.txt') as in_file:<br>                                 \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq_list = in_file.readlines()<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq_name = seq_list.pop(0)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq = seq_list.pop(0).rstrip().lower()<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for line in seq_list:<br> \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;seq += line.rstrip().lower()<br><br>     \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;with open('codons.txt') as in_file:<br>     \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codons_list = in_file.readlines()<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codons = {}  # Initialise the dictionary<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for count in range(0, len(codons_list), 2):<br>                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codons[codons_list[count].rstrip().lower()] =  codons_list[count+1].rstrip().lower()<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for count in range(0, len(seq), 3):<br> \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;codon = seq[count:count+3]<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if codon in codons:<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aa = codons[codon]<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;else:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;aa = '-'<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;print (aa, end=\" \")</b><br><br>\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 9\n",
    "\n",
    "---\n",
    "\n",
    "Write a program to test the sequence translation code. Modify the code so that it translates the sequence in 3 reading frames and prints them to a file. The sequences should be in fasta format with the frame number in the description line.\n",
    "\n",
    "The fasta file (<b>sequence.txt</b>) and codons file (<b>codons.txt</b>) needed to test the script can be downloaded from the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">Exercise Answers page</a>.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('sequence.txt') as in_file:\n",
    "    \n",
    "    seq_list = in_file.readlines()  # Read the file to a list\n",
    "    seq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "    seq = seq_list.pop(0)  # Initialise the sequence\n",
    "    seq = seq.rstrip()  # Remove the newline character\n",
    "    seq = seq.lower()  # Lower case the sequence\n",
    "\n",
    "    for line in seq_list:  #  Append the rest of the sequence to seq\n",
    "        seq += line.rstrip().lower()\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "    codons_list = codons_file.readlines()\n",
    "    codons = {}  # Initialise the dictionary\n",
    "    for count in range(0, len(codons_list), 2):  # Work through the list\n",
    "        key = codons_list[count].rstrip()   # Get the codon\n",
    "        key = key.lower()\n",
    "        value = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "        value = value.lower()\t\n",
    "        codons[key] = value  # Add to dictionary\n",
    "\n",
    "out_file = open(\"translated_seqs.fasta\", \"w\")\n",
    "for frame in range(0, 3):\n",
    "    out_file.write(\"> Frame \" + str(frame) + \"\\n\")\n",
    "\n",
    "    for count in range(frame, len(seq), 3):  # Work through the list\n",
    "        codon = seq[count:count+3]   # Get 3 nucleotides\n",
    "        if codon in codons:\n",
    "            aa = codons[codon]   # Get the associated aa\n",
    "        else:\n",
    "            aa = '-'\n",
    "        out_file.write(aa)\n",
    "    out_file.write(\"\\n\") # Sequence is finished so print newline for next sequence\n",
    "\n",
    "out_file.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='modules'></a>\n",
    "\n",
    "# Modules\n",
    "\n",
    "---\n",
    "\n",
    "Modules, or libraries, are common to most programming languages, including Perl, C++ and Java.\n",
    "\n",
    "Modules provide a set of code to provide particular functions that can be included in your own code. They are essentially programs with functions that can be called from your own program.\n",
    "\n",
    "As expected Python provides a large library of modules, including Biopython and PyCogent, and are imported with the \"import\" command.\n",
    "\n",
    "The most commonly used is probably the sys module which contains system-specific functionality. This is required for functions that are often included by default in other languages.\n",
    "\n",
    "---\n",
    "<a id='command_line_arg'></a>\n",
    "# Command Line Arguments\n",
    "\n",
    "---\n",
    "\n",
    "Often when writing a Python script you want to include the option to read values in to the script (command line arguments).\n",
    "\n",
    "In many languages this is a built in function but with Python you need to import the <b>sys</b> module.\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;import sys</b><br>\n",
    "\n",
    "Arguments can then be included when you run the script. For example, if the script is called test.py:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;python test.py arg1 arg2 arg3</b><br>\n",
    "\n",
    "To read them in to the script the <b>sys.argv</b> function is used:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;import sys<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a1 = sys.argv[1]<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a2 = sys.argv[2]<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a3 = sys.argv[3]</b><br>\n",
    "\n",
    "Now the variable a1 has the value of arg1 etc.\n",
    "\n",
    "It is possible to avoid having to write sys each time the function is called by explicitly importing that function.\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from sys import argv</b><br>\n",
    "\n",
    "Or to import all functions:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from sys import *</b><br>\n",
    "\n",
    "These options work for all modules and using either of the above the rest of the code is now simply:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;import sysa1 = argv[1]<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a2 = argv[2]<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a3 = argv[3]</b><br>\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 10\n",
    "\n",
    "---\n",
    "\n",
    "Write a script that reads two numbers as command line arguments, adds them together and prints the result.\n",
    " \n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '-f'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-1cd565c20ada>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mnum3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mn1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0mn2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '-f'"
     ]
    }
   ],
   "source": [
    "from sys import *\n",
    "\n",
    "def add(num1, num2): \n",
    "\tnum3 = num1 + num2\n",
    "\treturn num3 \n",
    "\n",
    "n1 = int(argv[1])\n",
    "n2 = int(argv[2])\n",
    "\n",
    "total = add(n1, n2)\n",
    "print (n1, \"+\", n2, \"=\", total)\n",
    "\n",
    "# Note that input needs to be converted to an integer.\n",
    "# Without this the 2 numbers would be treated as strings\n",
    "# and just appended together \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='string_methods'></a>\n",
    "\n",
    "# String Methods\n",
    "\n",
    "---\n",
    "\n",
    "Strings contain numerous built in methods enabling you to manipulate and interrogate them. Examples include:\n",
    "<a id='string_search'></a>\n",
    "## String Searches\n",
    "\n",
    "---\n",
    "\n",
    "<b>startswith(string)</b> – This will return true if the string starts with the the argument string, for example:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "protein = 'MEFTIKRDYFITQLNDT' \n",
    "\n",
    "# Does the protein start with methionine\n",
    "if protein.startswith('M'): \n",
    "    print ('Yes, the protein starts with methionine')\n",
    "           \n",
    "# Will print “Yes, the protein starts with methionine”\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "A substring can also be searched for within a string: \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "protein = 'MEFTIKRDYFITQLNDT'\n",
    "\n",
    "# Does the protein contain DYF\n",
    "if 'DYF' in protein: \n",
    "    print ('Yes, it contains the pattern DYF')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "<a id='string_split'></a>\n",
    "## Splitting Strings\n",
    "\n",
    "---\n",
    "\n",
    "Strings have an inbuilt split method that splits the string based on a delimiter and returns the split items in a list:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;str.split([delim[, maxsplit]])<br><br>\n",
    "\n",
    "If no arguments are provided the string wil be split on whitespace:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;' 1  2   3  '.split()<br>        \n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns ['1', '2', '3']</b><br><br>\n",
    "\n",
    "With delimiter:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'1XX2XX3'.split('XX')<br><br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns ['1', '2', '3']</b><br><br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "test_string =  '1 2 3'\n",
    "\n",
    "split_list = test_string.split()\n",
    "\n",
    "print (split_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Alternatively a delimiter can be used:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "test_string =  '1XX2XX3'   \n",
    "\n",
    "split_list = test_string.split('XX')\n",
    "\n",
    "print (split_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "If the optional <b>maxsplit</b> argument is provided it set the maximum number of splits in the string:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "test_string =  '1 2 3 4 5 6'\n",
    "\n",
    "split_list = test_string.split(\" \", 2)\n",
    "\n",
    "print (split_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "<a id='string_other'></a>\n",
    "## Other Methods\n",
    "\n",
    "---\n",
    "\n",
    "<b>lower() / upper()</b> - Returns a copy of the string converted to lower/upper case.\n",
    "\n",
    "<b>capitalize()</b> - Returns a copy of the string with only its first character capitalized.\n",
    "\n",
    "<b>replace(old, new[, count])</b>  Return a copy of the string with all occurrences of substring old replaced by new. If the optional argument count is given, only the first count occurrences are replaced.\n",
    "\n",
    "<b>replace(old, new[, count])</b> - Returns a copy of the string with all occurrences of substring old replaced by new. If the optional argument count is given, only the first count occurrences are replaced.\n",
    "\n",
    "There are many more.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace valine with tyrosine\n",
    "\n",
    "protein = \"vlspadktnv\"\n",
    "print(protein)\n",
    "\n",
    "new_protein = protein.replace(\"v\", \"y\")\n",
    "\n",
    "print(new_protein)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace more than one amino acid\n",
    "\n",
    "protein = \"vlspadktnv\"\n",
    "print(protein)\n",
    "\n",
    "new_protein = protein.replace(\"pad\", \"tyc\")\n",
    "\n",
    "print(new_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are many more.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 11\n",
    "\n",
    "---\n",
    "\n",
    "The file CAUH01000012.gff is a sequence annotation file in GFF format and is available on the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">Exercise Answers</a> page.\n",
    "\n",
    "a) Write a script that reads this file and prints out the lines that start with “CAUH01000012“, the ID for the sequence.\n",
    "\n",
    "b) Modify the file to print out the start and end position of the CDS regions.\n",
    "\n",
    "The columns in the GFF file are Sequence ID, Source, Type, Start, End, Score, Strand, Phase and Attributes. The CDS is identified in the third column and start and end in the fourth and fifth columns. The columns are all separated by tabs, which are whitespace.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAUH01000012\tBluGen\tgene\t4550\t5817\t.\t+\t.\tID=BGHDH14_bgh01634;Alias=bgh01634;Name=hydroxyisocaproate dehydrogenase;Ontology_term=GO:0008152,GO:0005488;inference=ab initio prediction:Eugene:3.5g:similar to RNA sequence, EST (same species):dbEST:GT062874.1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t4550\t4636\t.\t+\t.\tID=BGHDH14_bgh01634m1;Parent=BGHDH14_bgh01634;Name=BGHDH14_bgh01634m1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t4698\t4796\t.\t+\t.\tID=BGHDH14_bgh01634m2;Parent=BGHDH14_bgh01634;Name=BGHDH14_bgh01634m2\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t4852\t5565\t.\t+\t.\tID=BGHDH14_bgh01634m3;Parent=BGHDH14_bgh01634;Name=BGHDH14_bgh01634m3\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t5623\t5817\t.\t+\t.\tID=BGHDH14_bgh01634m4;Parent=BGHDH14_bgh01634;Name=BGHDH14_bgh01634m4\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t4550\t4636\t.\t+\t0\tParent=BGHDH14_bgh01634m1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t4698\t4796\t.\t+\t0\tParent=BGHDH14_bgh01634m2\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t4852\t5565\t.\t+\t0\tParent=BGHDH14_bgh01634m3\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t5623\t5715\t.\t+\t0\tParent=BGHDH14_bgh01634m4\t\n",
      "\n",
      "CAUH01000012\tBluGen\tgene\t7647\t8660\t.\t+\t.\tID=BGHDH14_bghG000012000001001;Alias=bghG000012000001001;Name=histone chaperone asf1;inference=ab initio prediction:Genemark:2.2a\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t7647\t7759\t.\t+\t.\tID=BGHDH14_bghG000012000001001m1;Parent=BGHDH14_bghG000012000001001;Name=BGHDH14_bghG000012000001001m1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t7854\t8216\t.\t+\t.\tID=BGHDH14_bghG000012000001001m2;Parent=BGHDH14_bghG000012000001001;Name=BGHDH14_bghG000012000001001m2\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t8267\t8660\t.\t+\t.\tID=BGHDH14_bghG000012000001001m3;Parent=BGHDH14_bghG000012000001001;Name=BGHDH14_bghG000012000001001m3\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t7647\t7759\t.\t+\t0\tParent=BGHDH14_bghG000012000001001m1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t7854\t8216\t.\t+\t0\tParent=BGHDH14_bghG000012000001001m2\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t8267\t8660\t.\t+\t0\tParent=BGHDH14_bghG000012000001001m3\t\n",
      "\n",
      "CAUH01000012\tBluGen\tgene\t17051\t17803\t.\t-\t.\tID=BGHDH14_bghG000012000002001;Alias=bghG000012000002001;Name=hypothetical protein;\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t17433\t17803\t.\t-\t.\tID=BGHDH14_bghG000012000002001m1;Parent=BGHDH14_bghG000012000002001;Name=BGHDH14_bghG000012000002001m1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tmRNA\t17051\t17370\t.\t-\t.\tID=BGHDH14_bghG000012000002001m2;Parent=BGHDH14_bghG000012000002001;Name=BGHDH14_bghG000012000002001m2\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t17433\t17709\t.\t-\t0\tParent=BGHDH14_bghG000012000002001m1\t\n",
      "\n",
      "CAUH01000012\tBluGen\tCDS\t17243\t17370\t.\t-\t0\tParent=BGHDH14_bghG000012000002001m2\t\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open('CAUH01000012.gff') as in_file:\n",
    "    for line in in_file: \n",
    "        if line.startswith('CAUH01000012'):\n",
    "            print (line) \n",
    "# \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start = 4550 and end = 4636\n",
      "Start = 4698 and end = 4796\n",
      "Start = 4852 and end = 5565\n",
      "Start = 5623 and end = 5715\n",
      "Start = 7647 and end = 7759\n",
      "Start = 7854 and end = 8216\n",
      "Start = 8267 and end = 8660\n",
      "Start = 17433 and end = 17709\n",
      "Start = 17243 and end = 17370\n"
     ]
    }
   ],
   "source": [
    "with open('CAUH01000012.gff') as in_file:\n",
    "    for line in in_file: \n",
    "        if line.startswith('CAUH01000012'):\n",
    "            values = line.split()\n",
    "            if values[2] == 'CDS':\t\n",
    "                print (\"Start =\", values[3], \"and end =\", values[4])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='functions'></a>\n",
    "\n",
    "# Functions\n",
    "\n",
    "---\n",
    "\n",
    "Often when writing a program there are times when you want to use the same piece of code multiple times. An example is a sequence translation program. What if you wanted to translate multiple sequence?\n",
    "\n",
    "This is where functions are used (or subroutines or methods, depending on the language).\n",
    "\n",
    "A function is a block of code that can be used multiple times to perform a particular piece of work.\n",
    "\n",
    "A function is defined by the keyword def followed by the function name. Note the brackets following the name, as functions can optionally take arguments.\n",
    "\n",
    "The structure of a function is like Python control methods, the function name ending in a colon and the scope of the function identified by indented code:\n",
    "\n",
    "<b>def hello():<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;print (\"Hello World!\")</b><br><br>\n",
    "    \n",
    "\n",
    "A function has to be declared in the program before it can be used.\n",
    "\n",
    "The function is called by simply using its name, and including arguments if required.\n",
    "\n",
    "The code within the function is then called:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def hello():\n",
    "\n",
    "    print (\"Hello World!\")\n",
    "       \n",
    "hello()\n",
    "hello()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "<a id='fun_pass_arg'></a>\n",
    "## Passing Arguments\n",
    "\n",
    "---\n",
    "\n",
    "Any number of arguments can be passed to a function:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def hello(name1, name2):\n",
    "    print (\"Hello\", name1, \"and\", name2)\n",
    "       \n",
    "\n",
    "hello(\"Lenka\", \"Ema\")\n",
    "                        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "<a id='fun_ret_val'></a>\n",
    "## Returning Values\n",
    "\n",
    "---\n",
    "\n",
    "Functions can also return data:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add(num1, num2):\n",
    "    num3 = num1 + num2\n",
    "    return num3\n",
    "\n",
    "\n",
    "num = add(3, 4)\n",
    "print (num)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Multiple values can be returned and assigned to variables when the function is called:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def calc(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 - num2\n",
    "\n",
    "    return a,b,c\n",
    "\n",
    "   \n",
    "d, e, f = calc(10, 5)\n",
    "print (d, e, f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Although the previous version actually returns a tuple it is possible to explicitly return one, or a list:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def calcTuple(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 - num2\n",
    "    \n",
    "    return (a,b,c)\n",
    "\n",
    "def calcList(num1, num2):\n",
    "    a = num1 + num2\n",
    "    b = num1 * num2\n",
    "    c = num1 - num2\n",
    "\n",
    "    return [a,b,c] \n",
    "\n",
    "t = calcTuple(10, 5)\n",
    "\n",
    "print (\"A tuple of values is returned:\")\n",
    "print (t)\n",
    "\n",
    "l = calcList(10, 5)\n",
    "\n",
    "print (\"A list of values is returned:\")\n",
    "print (l)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "## Exercise 12\n",
    "\n",
    "a) Write a script that includes a function that takes 2 numbers, multiplies them together and returns the result. The script should prompt the user for the numbers, pass them to the function and print out the result.\n",
    "\n",
    "b) Create a script that tests the example functions above, which return multiple values, a list and a tuple.\n",
    "\n",
    "Edit the script to demonstrate that it does actually return a list and a tuple.\n",
    "\n",
    "HINT: You can do this by assigning the returned values to a single variable, which will be the tuple, and then printing the index positions.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter first number: 1\n",
      "Enter second number: 2\n",
      "1 + 2 = 3\n"
     ]
    }
   ],
   "source": [
    "def add(num1, num2): \n",
    "\tnum3 = num1 + num2\n",
    "\treturn num3 \n",
    "\n",
    "n1 = int(input(\"Enter first number: \"))\n",
    "n2 = int(input(\"Enter second number: \"))\n",
    "\n",
    "total = add(n1, n2)\n",
    "print (n1, \"+\", n2, \"=\", total)\n",
    "\n",
    "\n",
    "# This code is obviously more verbose than it needs to be,\n",
    "# for instance there is no need for the \"total\" variable\n",
    "# and the print statement could be:\n",
    "#\n",
    "# print (n1, \"+\", n2, \"=\", add(n1, n2))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 50 5\n",
      "First value = 15\n",
      "Second value = 50\n",
      "Third value = 5\n",
      "First value = 15\n",
      "Second value = 50\n",
      "Third value = 5\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Calc function\n",
    "\n",
    "def calc(num1, num2):\n",
    "    \ta = num1 + num2\n",
    "    \tb = num1 * num2\n",
    "    \tc = num1 - num2\n",
    "\n",
    "    \treturn a,b,c \n",
    "\t\n",
    "d, e, f = calc(10, 5)\n",
    "print (d, e, f)\n",
    "\n",
    "\n",
    "# Tuple function\n",
    "\n",
    "def calcTuple(num1, num2):\n",
    "    \ta = num1 + num2\n",
    "    \tb = num1 * num2\n",
    "    \tc = num1 - num2\n",
    "\n",
    "    \treturn (a,b,c) \n",
    "\t\n",
    "tup = calcTuple(10, 5)\n",
    "print (\"First value =\", tup[0])\n",
    "print (\"Second value =\", tup[1])\n",
    "print (\"Third value =\", tup[2])\n",
    "\n",
    "\n",
    "# List function\n",
    "\n",
    "def calcList(num1, num2):\n",
    "    \ta = num1 + num2\n",
    "    \tb = num1 * num2\n",
    "    \tc = num1 - num2\n",
    "\n",
    "    \treturn (a,b,c) \n",
    "\t\n",
    "lis = calcList(10, 5)\n",
    "print (\"First value =\", lis[0])\n",
    "print (\"Second value =\", lis[1])\n",
    "print (\"Third value =\", lis[2])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='fun_pass_other'></a>\n",
    "## Passing Other Data as an Argument\n",
    "\n",
    "---\n",
    "\n",
    "Any data type, or object, can be passed to a Python function e.g. list, tuple, dictionary etc:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def calcSum(list1):\n",
    "    total = 0\n",
    "    for num in list1:\n",
    "        total += num\n",
    "\n",
    "    return total\n",
    "\n",
    "\n",
    "sumlist = (1, 2, 3, 4, 5)\n",
    "print (calcSum(sumlist))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "## Exercise 13\n",
    "\n",
    "---\n",
    "\n",
    "Create a script that contains a function that takes a dictionary as an argument and tests it.\n",
    "\n",
    "Edit the script further so that the function takes a list and dictionary as arguments and tests it. A suitable test would be for the list to contain the key values from the dictionary and then iterate the list in the function and print out the associated dictionary values.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Key is  ttt and value is F\n",
      "Key is  tta and value is L\n",
      "Key is  gga and value is G\n",
      "Key is ttt and value is F\n",
      "Key is tta and value is L\n",
      "Key is gga and value is G\n"
     ]
    }
   ],
   "source": [
    "# Dictionary function\n",
    "# Prints the keys and values of a dictionary\n",
    "\n",
    "def showDictionary(dic):\n",
    "\tfor k in dic.keys():\n",
    "\t\tprint (\"Key is \", k, \"and value is\", dic[k])\n",
    "\n",
    "\t\n",
    "codons = {'ttt':'F', 'tta':'L', 'gga':'G'}\n",
    "showDictionary(codons)\n",
    "\n",
    "# Dictionary and list function\n",
    "# Prints the keys and values of a dictionary using list of keys\n",
    "\n",
    "def showDictionary(lis, dic):\n",
    "\tfor k in lis:\n",
    "\t\tprint (\"Key is\", k, \"and value is\", dic[k])\n",
    "\n",
    "\n",
    "codon_list = ('ttt', 'tta', 'gga')\t\n",
    "codons = {'ttt':'F', 'tta':'L', 'gga':'G'}\n",
    "showDictionary(codon_list, codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "<a id='fun_pass_keyword'></a>\n",
    "## Passing Function Arguments by Keywords\n",
    "\n",
    "\n",
    "In the examples above, when multiple arguments are passed to a function, they are associated to parameters by their position. For example, with the following function definition:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;def calcValue(num1, num2, num3):</b><br>\n",
    "\n",
    "If it is called by:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;calcValue(5, 12, 16)</b><br>\n",
    "\n",
    "Then 5 would be assigned to num1, 12 to num2 and 16 to num3, as expected.\n",
    "\n",
    "However, arguments can also be passed by keywords and are then assigned to parameters by explicitly naming them. A simple calculation function could be:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n"
     ]
    }
   ],
   "source": [
    " \n",
    "def calcValue(num1, num2, num3):\n",
    "    total = num1 + num2 - num3\n",
    "   \n",
    "    return total\n",
    "\n",
    "\n",
    "ans = calcValue(5, 10, 6)\n",
    "\n",
    "print (ans)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Alternatively the arguments to the function can be explicitly named, and therefore in any order:\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def calcValue(num1, num2, num3):\n",
    "    total = num1 + num2 - num3\n",
    "   \n",
    "    return total\n",
    "\n",
    "ans = calcValue(num1=5, num2=10, num3=6)\n",
    "print (ans)\n",
    "\n",
    "ans = calcValue(num3=6, num2=10, num1=5)\n",
    "print (ans)\n",
    "\n",
    "ans = calcValue(num2=10, num1=5, num3=6)\n",
    "print (ans)\n",
    "\n",
    "# etc ...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "All versions will produce the same answer.\n",
    "\n",
    "An advantage is that you do not have to know in what order parameters are declared in the function.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 14\n",
    "\n",
    "---\n",
    "\n",
    "Create a script that tests the various ways to pass arguments to a function.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter first number: 1\n",
      "Enter second number: 2\n",
      "1 + 2 = 3\n"
     ]
    }
   ],
   "source": [
    "def add(num1, num2): \n",
    "\tnum3 = num1 + num2\n",
    "\treturn num3 \n",
    "\n",
    "n1 = int(input(\"Enter first number: \"))\n",
    "n2 = int(input(\"Enter second number: \"))\n",
    "\n",
    "total = add(n1, n2)\n",
    "print (n1, \"+\", n2, \"=\", total)\n",
    "\n",
    "\n",
    "# This code is obviously more verbose than it needs to be,\n",
    "# for instance there is no need for the \"total\" variable\n",
    "# and the print statement could be:\n",
    "#\n",
    "# print (n1, \"+\", n2, \"=\", add(n1, n2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<a id='fun_default_values'></a>\n",
    "\n",
    "## Default Values of Parameters\n",
    "\n",
    "---\n",
    "\n",
    "It is possible to set default values of parameters in the function definition:\n",
    " \n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;def calcValue(num1, num2=20, num3=8):</b><br>\n",
    "\n",
    "The parameters num1 and num2 now have default values so the function can be called with just a single value for num3, or with 2 values and num3 will be the default, or with 3 values and both defaults will be overwritten. Finally, one of the default values can be changed if required by specifying the defined name:\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def calcValue(num1, num2=20, num3=8):\n",
    "    total = num1 + num2 - num3\n",
    "\n",
    "    return total\n",
    "\n",
    "# Function can be called with just a single value for num\n",
    "ans = calcValue(6)\n",
    "print (\"Version 1: \", ans)\n",
    "\n",
    "# Called with 2 values and num3 will be the default\n",
    "ans = calcValue(6, 10)\n",
    "print (\"Version 2: \", ans)\n",
    "\n",
    "# Called  with 3 values and both defaults will be overwritten\n",
    "ans = calcValue(6, 10, 5)\n",
    "print (\"Version 3: \", ans)\n",
    "\n",
    "# Default value changed by specifying the defined name\n",
    "ans = calcValue(6, num3=4)\n",
    "print (\"Version 4: \", ans)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "NOTE: An example of the use of default values is with the Python open file command when opening a file for reading:\n",
    "\n",
    "<b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;f = open(‘test.txt’, ‘r’)</b><br>\n",
    "\n",
    "The ‘r’ is optional as it is a default value.\n",
    "\n",
    "---\n",
    "<a id='fun_var_params'></a>\n",
    "## Variable Number of Parameters\n",
    "\n",
    "---\n",
    "\n",
    "A function can take additional optional arguments by prefixing the last parameter with *.\n",
    "\n",
    "Optional arguments are then available in a tuple referenced by this parameter:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def calcValue(num1, num2, *morenums):\n",
    "    total = num1 + num2\n",
    "    for n in morenums:\n",
    "        total += n\n",
    "\n",
    "    return total\n",
    "\n",
    "  \n",
    "ans = calcValue(2, 4,  5, 12, 15)\n",
    "print (ans)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In this case 2 would be assigned to num1, 4 to num2 and tuple created from the rest (5, 12, 15).\n",
    " \n",
    "If ** is used the arguments are then converted in to a dictionary.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def testFunc (val1, val2, **morevals):\n",
    "    print (\"Val1 is\", val1)\n",
    "    print (\"Val2 is\", val2)\n",
    "\n",
    "    print (\"The dictionary:\")\n",
    "\n",
    "    for k in morevals.keys():\n",
    "        print (\"Key is\", k, \"Value is\", morevals[k])\n",
    "\n",
    "testFunc(\"First\", \"Second\", Third= 3, Fourth= 4, Fifth= 5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "NOTE: The keys in the dictionary section of the function call do not have quotes.\n",
    "\n",
    "---\n",
    "\n",
    "## Exercise 15\n",
    "\n",
    "---\n",
    "\n",
    "Create a script that tests the examples given above for default parameters and variable numbers of parameters.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "8\n",
      "11\n",
      "22\n",
      "38\n",
      "Val1 is First\n",
      "Val2 is Second\n",
      "The dictionary\n",
      "Key is Third Value is 3\n",
      "Key is Fourth Value is 4\n",
      "Key is Fifth Value is 5\n"
     ]
    }
   ],
   "source": [
    "# Default parameters\n",
    "\n",
    "def calcValue(num1, num2=20, num3=8):\n",
    "\n",
    "\ttotal = num1 + num2 - num3\n",
    "\n",
    "\treturn total\n",
    "\n",
    "\n",
    "ans = calcValue(6)\n",
    "\n",
    "print (ans)\n",
    "\n",
    "\n",
    "# Call it with 2 numbers so only num3 uses the default\n",
    "\n",
    "ans = calcValue(6, 10)\n",
    "\n",
    "print (ans)\n",
    "\n",
    "\n",
    "# Can still call it with 3 different numbers\n",
    "\n",
    "ans = calcValue(6, 10, 5)\n",
    "\n",
    "print (ans)\n",
    "\n",
    "\n",
    "# Can also change just one of the defaults\n",
    "\n",
    "ans = calcValue(6, num3=4)\n",
    "\n",
    "print (ans)\n",
    "\n",
    "\n",
    "# Variable number of parameters in a list\n",
    "\n",
    "def calcValue(num1, num2, *morenums):\n",
    "\n",
    "\ttotal = num1 + num2\n",
    "\tfor n in morenums:\n",
    "\t\ttotal += n\n",
    "\n",
    "\treturn total\n",
    "\n",
    "\n",
    "ans = calcValue(2, 4,  5, 12, 15)\n",
    "print (ans)\n",
    "\t\t\n",
    "\n",
    "# Variable number of parameters in a dictionary\n",
    "\n",
    "def testFunc (val1, val2, **morevals):\n",
    "\n",
    "\tprint (\"Val1 is\", val1)\n",
    "\tprint (\"Val2 is\", val2)\n",
    "\n",
    "\tprint (\"The dictionary\")\n",
    "\n",
    "\tfor k in morevals.keys():\n",
    "\t\tprint (\"Key is\", k, \"Value is\", morevals[k])\n",
    "\n",
    "\n",
    "\n",
    "testFunc(\"First\", \"Second\", Third=3, Fourth=4, Fifth=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Exercise 16\n",
    "\n",
    "---\n",
    "\n",
    "In the Example Program - Sequence Translation part of the tutorial you were provided with the code to translate a nucleotide sequence into amino acids, which you tested in exercise 8.\n",
    "\n",
    "The fasta file (<b>sequence.txt</b>) and codons file (<b>codons.txt</b>) needed to test the script can be downloaded from the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">Exercise Answers</a> page.\n",
    "\n",
    "Modify the code so that it translates the sequence in all 6 reading frames. You should include a function that carries out the translation and call it 6 times.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward Strand\n",
      "Frame 1 :\n",
      " MDLEKNYPTPRTSRTGHGGVNQLGGVFVNGRPLPDVVRQRIVELAHQGVRPCDISRQLRVSHGCVSKILGRYYETGSIKPGVIGGSKPKVATPKVVEKIAEYKRQNPTMFAWEIRGRLLAERVCDNDTVPSVSSINRIIRTKVQQPPNQPVPASSHSIVSTGSVTQVSSVSTDSAGSSYSISGILGITSPSADTNKRKRDEGIQESPVPNGYSLPGRDFLRKQMRGDLFTQQQLEVLDRVFERQHYSDIFTTTEPIKPEQTTEYSAMASLAGGLDDMKANLASPTPADIGSSVPGPQSYPIVTGRDLASTTLPGYPPHVPPAGQGSYSAPTLTGMVPGSEFSGSPYSHPQYSSYNDSWRFPNPGLLGSPYYYSAAARGAAPPAAATAYDRH-\n",
      "Frame 2 :\n",
      " WI*RKIIRLLGPAGQDMEE*ISLGGFL*MDGHSRM*SARG*WNLLIKVSGPATSPGSFGSAMVVSAKFLAGIMRQEASSLG*LEDPNQRSPHPKWWKKSLSINAKIPPCLPGRSGAGCWQSGCVTMTPCLASVPSTGSSGQKYSSHPTNQSQLPVTA*CPLAP*RRCPR*ARIRPARRTPSAASWASRPPAPTPTSARETKVFRSLRCRTATRFRAETSSGSRCGETCSHSSSWRCWTACLRGSTTQTSSPPQSPSSPSRPQSIQPWPRWLVGWTT*RPIWPAPPLLTSGAVCQARSPTPL*QAVTWRARPSPGTLHTSPPLDRAATQHRR*QGWCLGVSFPGVPTATLSIPRTTTPGGSPTRGCLAPPTIIALPPEEPPHLQPPLPMTVT-\n",
      "Frame 3 :\n",
      " GFREKLSDSSDQQDRTWRSESAWGGFCEWTATPGCSPPEDSGTCSSRCQALRHLQAASGQPWLCQQNSWQVL*DRKHQAWGNWRIQTKGRHTQSGGKNR*V*TPKSHHVCLGDQGPAAGRAGV*Q*HRA*RQFHQQDHPDKSTAATQPTSPSFQSQHSVHWLRDAGVLGEHGFGRLVVLHQRHPGHHVPQRRHQQAQERRRYSGVSGAERLLASGQRLPPEADAGRLVHTAAAGGAGPRV*EAALLRHLHHHRAHQARADHRVFSHGLAGWWAGRHEGQSGQPHPC*HREQCARPAVLPHCDRP*LGEHDPPRVPSTRPPRWTGQLLSTDADRDGAWE*VFRESLQPPSVFLVQRLLEVPQPGAAWLPLLL*RCRPRSRPTCSRHCL*PSL\n",
      "\n",
      "Reverse Strand\n",
      "Frame 1 :\n",
      " Q*RS*AVAAAGGAAPRAAAL***GEPSSPGLGNLQESLYEEY*GWL*GLPENSLPGTIPVSVGAE*LPCPAGGTCGGYPGRVVLAKSRPVTMG*DCGPGTLLPMSAGVGLARLAFMSSSPPASEAMAEYSVVCSGLMGSVVVKMSE*CCLSNTRSSTSSCCCVNKSPRICFRRKSLPGSE*PFGTGDS*IPSSLLRLLVSALGDVMPRMPLMEYDEPAESVLTEDTCVTEPVDTML*LEAGTGWLGGCCTFVRMILLMELTLGTVSLSHTRSASSRPLISQANMVGFWRLYSAIFSTTLGVATFGLDPPITPGLMLPVS*YLPRILLTQPWLTRSCLEMSQGLTP**ASSTILWRTTSGSGRPFTKTPPS*FTPPCPVLLVRGVG*FFSKS-\n",
      "Frame 2 :\n",
      " SDGHRQWRLQVGRLLGRQRYNSRGSQAAPGWGTSRSRCTRNTEGGCRDSRKTHSQAPSLSASVLSSCPVQRGGRVEGTRGGSCSPSHGLSQWGRTAGLAHCSRCQQGWGWPDWPSCRPAHQPARPWLNTLWSARA*WALWW*RCLSSAASQTRGPAPPAAAV*TSLPASASGGSLCPEASSRSAPETPEYLRLSCACWCRRWGT*CPGCR*WSTTSRPNPCSPRTPASRSQWTLCCDWKLGLVGWVAAVLLSG*SC*WN*R*ARCHCHTPALPAAGP*SPRQTWWDFGVYTQRFFPPLWVWRPLVWILQLPQA*CFLSHNTCQEFC*HNHG*PEAAWRCRRA*HLDEQVPLSSGGLHPGVAVHSQKPPQADSLLHVLSCWSEESDNFSLNP-\n",
      "Frame 3 :\n",
      " VTVIGSGGCRWGGSSGGSAIIVGGAKQPRVGEPPGVVVRGILRVAVGTPGKLTPRHHPCQRRC*VAALSSGGDVWRVPGEGRARQVTACHNGVGLRAWHTAPDVSRGGAGQIGLHVVQPTSQRGHG*ILCGLLGLDGLCGGEDV*VVLPLKHAVQHLQLLLCEQVSPHLLPEEVSARKRVAVRHRRLLNTFVSLALVGVGAGGRDAQDAADGVRRAGRIRAHRGHLRHGASGHYAVTGSWDWLVGWLLYFCPDDPVDGTDARHGVIVTHPLCQQPAPDLPGKHGGILAFILSDFFHHFGCGDLWFGSSNYPRLDASCLIIPAKNFADTTMADPKLPGDVAGPDTLMSKFHYPLADYIREWPSIHKNPPKLIHSSMSCPAGPRSRIIFL*IH\n"
     ]
    }
   ],
   "source": [
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "\n",
    "def translate(seq):\n",
    "    aaseq = \"\"\n",
    "    for count in range(0, len(seq), 3):  # Work through the list\n",
    "        codon = seq[count:count+3]   # Get 3 nucleotides\n",
    "        if codon in codons:\n",
    "            aa = codons[codon]   # Get the associated aa\n",
    "        else:\n",
    "            aa = '-'\n",
    "\n",
    "        aaseq += aa\n",
    "\n",
    "    return aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "    codons_list = codons_file.readlines()\n",
    "\n",
    "    codons = {}  # Initialise the dictionary\n",
    "    for count in range(0, len(codons_list), 2):  # Work through the list\n",
    "        key = codons_list[count].rstrip()   # Get the codon\n",
    "        key = key.upper()\n",
    "        value = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "        value = value.upper()\t\n",
    "        codons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('sequence.txt') as in_file:\n",
    "\n",
    "    seq_list = in_file.readlines()  # Read the file to a list\n",
    "    seq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "    seq = seq_list.pop(0)  # Initialise the sequence\n",
    "    seq = seq.rstrip()  # Remove the newline character\n",
    "\n",
    "    for line in seq_list:  #  Append the rest of the sequence to seq\n",
    "        seq += line.rstrip()\n",
    "    seq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "    for i in range(1, 3):\n",
    "        if i == 1:\n",
    "            print (\"Forward Strand\")\n",
    "        else:\n",
    "            print (\"\\nReverse Strand\")\n",
    "        for j in range(1, 4):\n",
    "            aaseq = translate(seq[j-1:])\n",
    "\n",
    "            print (\"Frame\", j, \":\\n\", aaseq)\n",
    "\n",
    "        # Reverse the sequence\n",
    "        revseq = \"\"\n",
    "        seq = seq.upper()\n",
    "        if i == 1:\n",
    "            seq = seq[::-1]\n",
    "            for base in seq:\n",
    "                if base in trans:\n",
    "                    revseq += trans[base]\n",
    "                else:\n",
    "                    revseq += \"N\"\n",
    "            seq = revseq\n",
    "            \n",
    "\n",
    "            \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Exercise 17\n",
    "\n",
    "---\n",
    "\n",
    "Modify your code from the previous exercise so that the function translates only a section of the sequence. It should also only translate the sequence in a single frame, defined as a function parameter.\n",
    "\n",
    "\n",
    "For example, the function could take 4 arguments:\n",
    "\n",
    "    Sequence\n",
    "    Start\n",
    "    End\n",
    "    Sequence strand (1 or -1)\n",
    "\n",
    "Remember, for the minus strand you will need to complement the sequence.\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LHPGVAVHSQKPPQADSLLHVLSCWSEESDNFS\n"
     ]
    }
   ],
   "source": [
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "\n",
    "def translate(seq, start, end, strand):\n",
    "\n",
    "    seq = seq[start:end]\n",
    "\n",
    "    if strand == -1:\n",
    "        # Reverse the sequence\n",
    "        revseq = \"\"\n",
    "\n",
    "        seq = seq[::-1]\n",
    "        for base in seq:\n",
    "            if base in trans:\n",
    "                revseq += trans[base]\n",
    "            else:\n",
    "                revseq += \"N\"\n",
    "        seq = revseq\n",
    "\n",
    "    aaseq = \"\"\n",
    "    for count in range(0, len(seq), 3):  # Work through the list\n",
    "        codon = seq[count:count+3]   # Get 3 nucleotides\n",
    "        if codon in codons:\n",
    "            aa = codons[codon]   # Get the associated aa\n",
    "        else:\n",
    "            aa = '-'\n",
    "\n",
    "        aaseq += aa\n",
    "\n",
    "    return aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "    codons_list = codons_file.readlines()\n",
    "\n",
    "    codons = {}  # Initialise the dictionary\n",
    "    for count in range(0, len(codons_list), 2):  # Work through the list\n",
    "        key = codons_list[count].rstrip()   # Get the codon\n",
    "        key = key.upper()\n",
    "        value = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "        value = value.upper()\t\n",
    "        codons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('sequence.txt') as in_file:\n",
    "\n",
    "    seq_list = in_file.readlines()  # Read the file to a list\n",
    "    seq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "    seq = seq_list.pop(0)  # Initialise the sequence\n",
    "    seq = seq.rstrip()  # Remove the newline character\n",
    "\n",
    "    for line in seq_list:  #  Append the rest of the sequence to seq\n",
    "        seq += line.rstrip()\n",
    "    seq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "    aaseq = translate(seq, 10, 109, -1)\n",
    "\n",
    "    print (aaseq)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Exercise 18\n",
    "\n",
    "---\n",
    "\n",
    "Modify your code from the previous exercise so that the function extracts  multiple sections of the sequence and translates them. The previous example would only translate a single exon gene but this version should translate a multi exon gene.\n",
    "\n",
    "The number of exons will vary so your function will have to take a single list of start and end positions.\n",
    "\n",
    "To test this version of the function you can use a different and much larger sequence (<b>blumeria_seq.fasta</b>), available from the <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">Exercise Answers</a> page. This is a contig annotation from Blumeria graminis and the genes in the sequence are annotated so you can use those positions to test your function. There are 3 genes, 2 on the forward strand and 1 on the reverse:\n",
    "\n",
    "Gene 1\n",
    "\n",
    "Strand +\n",
    "Exon 1     Start     4550     End     4636           \n",
    "Exon 2     Start     4698     End    4796       \n",
    "Exon 3     Start     4852     End    5565   \n",
    "Exon 4     Start     5623     End     5715   \n",
    "\n",
    "Gene 2\n",
    "\n",
    "Strand +\n",
    "Exon 1     Start      7647       End      7759   \n",
    "Exon 2     Start      7854      End      8216   \n",
    "Exon 3     Start      8267      End      8660   \n",
    "\n",
    "Gene 3\n",
    "\n",
    "Strand –\n",
    "Exon 1     Start    17433    End      17709\n",
    "Exon 2     Start      17243  End      17370       \n",
    "\n",
    "You could also change the function to use arguments with default values. For example, a default could be “strand=1”, so the function assumes the positive strand.\n",
    "\n",
    "NOTE: The correct translations are:\n",
    "\n",
    "Gene 1 Translation: MSLPTVLLVYKIIHTVDEWNDLSRIATLKVFNGNSRQDFLKSCEDGEFNEVTIIYYSNESVRLIGQLDQELLFKLPPSIRYICHHGAGYDSIDIAACTERSIQVSHTPIAVNKSTADMTLFLILGALRRIHVPYLAVRAGQWRGSTQLGHDPENKLLGILGMGGIGQEVAKRAKAFGMKVQYHNRSRLVPELEQGADYVSFEHLIKTSDILSLNCSLNKATIGIIGKDELAQMKQGVIVVNTARGKLIDEFALVKALESGKVFSAGLDVFEKEPSIDDTLLSSPNVVLTPHIGTATVETQRAMELLVLQNIENALKTDALVTQVQEQIQA*\n",
    "Gene 2 Translation: MSVVSLLGVNVLQNPARFGDPYEFEITFECLETLQKGTYIVGHKLTPVFNKLIFNSNDHDQELDSLLVGPIPVGVNKFIFVADPPDTNKIPDAEILGVTVILLTCAYDGREFVRVGYYVNNEYDSDELNTDPPAKPILEKVRRNILAEKPRVTRFAIKWDSDDSAPPLYPPEQPEADLVADGEEYGAEEAEDEEEEESADGPEVPADPDVMIEDSEVAGAMVETVKATEEESDAGSEDLEAESSGSEEDEIEEDEEREDEPEEAMDLDGAGKPNGATSSTHNTDTTMAH*\n",
    "Gene 3 Translation: MWSLQYLALLVFVSRAAANYQHWDIDSAVNCQNNIWTSNYLKQVRLRYCNHPPSPQSLVDISRHPNSQLHSNRSTNIYFLSIPRPGPDGELGDDVSNYYMLVDADCNYYSVVGLNARYAQGFVLNSQTVPCRLA*\n",
    "\n",
    "\n",
    "(Answers to all exercises are available <a href=\"http://teaching.bc.ic.ac.uk/msc/ipython-files/exercises.html\">here</a>.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene 1 Translation: MSLPTVLLVYKIIHTVDEWNDLSRIATLKVFNGNSRQDFLKSCEDGEFNEVTIIYYSNESVRLIGQLDQELLFKLPPSIRYICHHGAGYDSIDIAACTERSIQVSHTPIAVNKSTADMTLFLILGALRRIHVPYLAVRAGQWRGSTQLGHDPENKLLGILGMGGIGQEVAKRAKAFGMKVQYHNRSRLVPELEQGADYVSFEHLIKTSDILSLNCSLNKATIGIIGKDELAQMKQGVIVVNTARGKLIDEFALVKALESGKVFSAGLDVFEKEPSIDDTLLSSPNVVLTPHIGTATVETQRAMELLVLQNIENALKTDALVTQVQEQIQA*\n",
      "Gene 2 Translation: MSVVSLLGVNVLQNPARFGDPYEFEITFECLETLQKGTYIVGHKLTPVFNKLIFNSNDHDQELDSLLVGPIPVGVNKFIFVADPPDTNKIPDAEILGVTVILLTCAYDGREFVRVGYYVNNEYDSDELNTDPPAKPILEKVRRNILAEKPRVTRFAIKWDSDDSAPPLYPPEQPEADLVADGEEYGAEEAEDEEEEESADGPEVPADPDVMIEDSEVAGAMVETVKATEEESDAGSEDLEAESSGSEEDEIEEDEEREDEPEEAMDLDGAGKPNGATSSTHNTDTTMAH*\n",
      "Gene 3 Translation: MWSLQYLALLVFVSRAAANYQHWDIDSAVNCQNNIWTSNYLKQVRLRYCNHPPSPQSLVDISRHPNSQLHSNRSTNIYFLSIPRPGPDGELGDDVSNYYMLVDADCNYYSVVGLNARYAQGFVLNSQTVPCRLA*\n"
     ]
    }
   ],
   "source": [
    "# Dictionary for the reverse complementing of the nucleic acids\n",
    "\n",
    "trans = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "\n",
    "def translate(seq, strand, *pos):\n",
    "\n",
    "    code_seq = \"\"\n",
    "    for p in range(0, len(pos), 2):\n",
    "        start = pos[p]\n",
    "        end = pos[p+1]\n",
    "\n",
    "        ex_seq = seq[start-1:end]\n",
    "\n",
    "        if strand == -1:\n",
    "            # Reverse the sequence\n",
    "            revseq = \"\"\n",
    "\n",
    "            ex_seq = ex_seq[::-1]\n",
    "            for base in ex_seq:\n",
    "                if base in trans:\n",
    "                    revseq += trans[base]\n",
    "                else:\n",
    "                    revseq += \"N\"\n",
    "            ex_seq = revseq\n",
    "\n",
    "        code_seq += ex_seq\n",
    "\n",
    "    aaseq = \"\"\n",
    "    for count in range(0, len(code_seq), 3):  # Work through the list\n",
    "        codon = code_seq[count:count+3]   # Get 3 nucleotides\n",
    "        if codon in codons:\n",
    "            aa = codons[codon]   # Get the associated aa\n",
    "        else:\n",
    "            aa = '-'\n",
    "\n",
    "        aaseq += aa\n",
    "\n",
    "    return aaseq\n",
    "\n",
    "\n",
    "with open('codons.txt') as codons_file:\n",
    "    codons_list = codons_file.readlines()\n",
    "\n",
    "    codons = {}  # Initialise the dictionary\n",
    "    for count in range(0, len(codons_list), 2):  # Work through the list\n",
    "        key = codons_list[count].rstrip()   # Get the codon\n",
    "        key = key.upper()\n",
    "        value = codons_list[count+1].rstrip()  # Get the aa, on next line\n",
    "        value = value.upper()\t\n",
    "        codons[key] = value  # Add to dictionary\n",
    "\n",
    "with open('blumeria_seq.fasta') as in_file:\n",
    "\n",
    "    seq_list = in_file.readlines()  # Read the file to a list\n",
    "    seq_name = seq_list.pop(0)  # Remove the sequence name\n",
    "    seq = seq_list.pop(0)  # Initialise the sequence\n",
    "    seq = seq.rstrip()  # Remove the newline character\n",
    "\n",
    "    for line in seq_list:  #  Append the rest of the sequence to seq\n",
    "        seq += line.rstrip()\n",
    "    seq = seq.upper()  # Upper case the sequence\n",
    "\n",
    "    aaseq = translate(seq, 1, 4550, 4636, 4698, 4796, 4852, 5565, 5623, 5715)\n",
    "\n",
    "    print (\"Gene 1 Translation:\", aaseq)\n",
    "\n",
    "    aaseq = translate(seq, 1, 7647, 7759, 7854, 8216, 8267, 8660)\n",
    "\n",
    "    print (\"Gene 2 Translation:\", aaseq)\n",
    "\n",
    "    aaseq = translate(seq, -1, 17433, 17709, 17243, 17370)\n",
    "\n",
    "    print (\"Gene 3 Translation:\", aaseq)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<b>The third part of the Python tutorial is available <a href=\"PythonTutorial_Pt3.ipynb\">here</a>.</b><br><br>\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
